---
title: "Notes in Meta UM Proj"
author: "Phoebe Fei"
date: '2022-05-31'
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---

```{css style_settings, echo = FALSE}
blockquote { 
    font-size: 14px; 
    font-style: italic;
}
h1{
    font-size: 22px;
}
h2{
    font-size: 20px;
}
h3{
    font-size: 18px;
}
h4{
    font-size: 16px;
}

```

# Fastq to uBAM

## Notes:

Each sample is pair-end, and stored in /athena/nixonlab/scratch/projects/HERV_um_mets.git/raw. There are 5 fastq.gz files in each folder. *S.._R[1|2]_001.fastq.gz are the 2 files to use for downstream analysis.

```bach
nano qccheck.sh 

#! /bin/bash -l
#SBATCH --partition=panda
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --job-name=QC
#SBATCH --time=15:00:00 #HH/MM/SS
#SBATCH --mem=10G
spack load fastqc
cd /athena/nixonlab/scratch/projects/HERV_um_mets.git/raw
for d in */ ; do
    echo "$d"
    cd $d
    for f in *.fastq.gz; do
      fastqc $f -o /athena/nixonlab/scratch/tof4003/umprojscripts
    done
    cd ..
done 
cd /athena/nixonlab/scratch/tof4003/umprojscripts
spack load -r py-multiqc
multiqc .
```


### Now that there are samples, running DESeq2 on the Datasets

Sample names, organized by PD/SD
```{r}
samplenames <- c(
  #SD
  "SK_MEL_1316A_T"	,"SK_MEL_1313A_T","SK_MEL_1313B_T","SK_MEL_1306A_T","SK_MEL_1306B_T","SK_MEL_1331A_T","SK_MEL_1331B_T",	
  #PD
  "SK_MEL_1299A_T","SK_MEL_1299B_T","SK_MEL_1058B_T",	"SK_MEL_1330A_T","SK_MEL_1330B_T","SK_MEL_1328A_T",
  #PD (and has timepoint post-pd)
  "SK_MEL_1327A_T","SK_MEL_1327B_T")
```
Interested HERVs

```{r}
herv_profile <- c('ERVLE_17p11.2c','LTR46_Xq11.1','HERVH_6q24.1b','HERVH_3q21.3a','MER4B_3q21.2','HERVH_10q23.32','HML3_19q13.2','HERVI_9q22.1','MER4_1p21.3','HML3_3q21.2','ERVLE_5q13.2a','HERV3_19q13.42a','HERVE_Xp11.23','ERV316A3_6p21.33c','MER101_19q13.2c','ERVLE_2p13.2a','MER4B_12q22')

```

### Genes

#### Import readcounts and assign sample names
```{r}
genedirect <- "/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/align_multi/"
if(exists("bindtable")){
  rm("bindtable")}
for (i in 1:length(samplenames)){
  filedirect <- paste0(genedirect, samplenames[i],"/ReadsPerGene.out.tab")
  if(exists("bindtable")){
    test <- read.table(filedirect, header = FALSE, row.names = 1)
    #take out the summary rows
    sample.lane <- data.frame(test[-c(1:4),3], row.names = rownames(test)[-c(1:4)])
    bindtable <- cbind(bindtable,sample.lane[rownames(bindtable),])
  } else {
    test <- read.table(filedirect, header = FALSE, row.names = 1)
    #take out the summary rows
    bindtable <- data.frame(test[-c(1:4),3], row.names = rownames(test)[-c(1:4)])
  }
}
colnames(bindtable) <- samplenames
```

#### HERV counts


```{r, eval = FALSE}
devtools::install_github("nixonlab/scopetools")
```

```{r}
library(scopetools)
direct <- "/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/telescope/"
#create the directory path to reports.tsv
filedirect <- vector(length = length(samplenames))
for (i in 1:length(samplenames)){
  repo <- paste0(direct,samplenames[i], "/report.tsv")
  filedirect[i] <- repo
}
#make sample names as colnames
msk_report <- load_telescope_reports(files = filedirect, colnames = samplenames)

#take out no features 
msk_report <- msk_report[which(rownames(msk_report) != "__no_feature"),]
```

#### Metadata

```{r}
sample_meta <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/sample_info.csv")
#make sample name row name
row.names(sample_meta) <- make.names(sample_meta$SK.MEL)
#reorder so that the rowname oder matches the original counttable
sample_meta <- sample_meta[colnames(bindtable),]
```

#### Notes:

[RECIST Classification](https://www.cibmtr.org/DataManagement/TrainingReference/Manuals/DataManagement/Documents/appendix-n.pdf):

#### Response Evaluation Crteria in Solid Tumors

*SD* = Stable Disease (NR/SD) - neither sufficeint shrinkage to qualify for partial response nor sufficient increase to qualify for progressive disease (PD)

*PD* = Progressive Disease - A 20% or greater increase in the sum of the longest diameter of measured lesions.

#### Gene + HERV to Plot PCA
```{r}
bind_all <- rbind(bindtable[,samplenames], msk_report[,samplenames])
```

#### Primary Mets Data from TCGA

```{r}
load("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/primary_UM/analysis2/04-count_data.Rdata")
#data is the counts_herv
```


### Libraries
```{r, message=FALSE}
library(stringr)
library(DESeq2)
library(vsn)
library(ggplot2)
library(apeglm)
library(dplyr)
library(magrittr)
library(pheatmap)
library(EnhancedVolcano)
library(viridis)
library(biomaRt)
library(ggplot2)
```

## Functions

### Function to subset Metadata
```{r}
submeta <- function(status = "primary", met_meta, count_table){
  meta_sub <- data.frame(matrix(nrow = ncol(count_table), ncol = 2))
  rownames(meta_sub)<- colnames(count_table)
  colnames(meta_sub)<- c("BioSample", "status")
  meta_sub$BioSample <- colnames(count_table)
  meta_sub$status <- rep(status, nrow(meta_sub))
  m_um_sub <- met_meta[,"SK.MEL",drop = FALSE]
  colnames(m_um_sub) <- c("BioSample")
  m_um_sub$status <- rep("metastatic",nrow(m_um_sub))
  um_sub <- rbind(meta_sub,m_um_sub)
  return(um_sub)
}


```
### Result Table & HeatMap Table Generation

Result Table
```{r}
restable <- function(min_fc = 2, max_q = 0.05, n_top_genes = 20, deseq){
  MIN_L2FC=log2(min_fc)
  res_table <- results(deseq, independentFiltering = TRUE) %>% as.data.frame(.) %>% dplyr::select(everything()) %>%
    arrange(padj) %>%
    filter(log2FoldChange >= min_fc | log2FoldChange <= -min_fc) 
  print(res_table[1:n_top_genes, ])
  return(res_table)
}
```

Heatmap Table

```{r}
heattable <- function(deseq, herv_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 2, max_q = 0.05 ){
  #rlog scaled
  deseq.rlog <- varianceStabilizingTransformation(deseq, blind = TRUE)
  #select padj
  seg.herv <- subset(herv_res, padj < max_q & (log2FoldChange > log2(min_fc) | log2FoldChange < -log2(min_fc)))
  #Select the HREVs/Genes
  herv_names  <- str_extract(rownames(seg.herv), select_string)
  seg.herv.1 <- seg.herv[(rownames(seg.herv) %in%  herv_names),]
  rlog.herv <- deseq.rlog[rownames(seg.herv.1),] %>% assay
  return(rlog.herv)
}

```



### DESeq Objects

#### HERV+Gene - PCA Plotting

```{r}
#create the object
#RECIST colinear with Patient
DESeq.all <- DESeqDataSetFromMatrix(countData = as.matrix(bindtable),
                                   colData = sample_meta,
                                   design = ~ RECIST+Timepoint) 

```

Mean SD Plot + PCA
```{r}
par(mar=c(11, 8.1, 4.1, 2.1))
#size factor
DESeq.all <- estimateSizeFactors(DESeq.all)
#normalize & log-sized
DESeq.all_norm <- counts(DESeq.all, normalized = TRUE)
boxplot(log2(DESeq.all_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.all,"log.norm.counts") <- log2(DESeq.all_norm+1)

#SD and PCA plot
msd.all <- vsn::meanSdPlot(DESeq.all@assays@data@listData$log.norm.counts, ranks=FALSE, plot = FALSE)
msd.all$gg + ggtitle("Sequencing depth normalized log2(read counts)") + ylab("standard deviation")

#PCA
DESeq.all.trans <-DESeqTransform(DESeq.all)
plotPCA(DESeq.all.trans,intgroup = "Patient",returnData = FALSE) + ggtitle("All PCA Plot Patient")
plotPCA(DESeq.all.trans,intgroup = "RECIST",returnData = FALSE) + ggtitle("All PCA Plot RECIST")
plotPCA(DESeq.all.trans,intgroup = "Timepoint",returnData = FALSE) + ggtitle("All PCA Plot Timepoint")
plotPCA(DESeq.all.trans,intgroup = "RECIST_3m",returnData = FALSE) + ggtitle("All PCA Plot RECIST_3m")
```
Looks like only RECIST accounts for some variances in PC2. Will adjust for timepoint still, by only taking the pre-treatment group.


#### Take out the pre-treatment subgroup

```{r}
meta_sub <- sample_meta[which(sample_meta$Timepoint == "Pre-tx"),]
gene_table <- bindtable[,rownames(meta_sub)]
msk_sub <- msk_report[,rownames(meta_sub)]
```



#### Generate Logged Normalized Count Table for Cybersortx (All Samples)

*Reorganize so that samples separated to first SD then PD, within those pre-tx, on-tx post-PD)*

```{r}
cyber_sample <-  c(
  #SD + Pre-tx
  "SK_MEL_1316A_T","SK_MEL_1313A_T","SK_MEL_1306A_T","SK_MEL_1331A_T",
  #SD + On-tx
  "SK_MEL_1313B_T","SK_MEL_1306B_T","SK_MEL_1331B_T",
  #PD + Pre-tx
  "SK_MEL_1299A_T","SK_MEL_1058B_T","SK_MEL_1327A_T",	"SK_MEL_1330A_T","SK_MEL_1328A_T",
  #PD + On-tx
  "SK_MEL_1299B_T","SK_MEL_1330B_T",
  #PD + Post-PD
  "SK_MEL_1327B_T")
all_genes <- bindtable[,cyber_sample]
cyber_meta <- sample_meta[cyber_sample,]
```

```{r}
#create the object
#Include RECIST in design
DESeq.gene.c <- DESeqDataSetFromMatrix(countData = as.matrix(all_genes),
                                   colData = cyber_meta,
                                   design = ~ RECIST) 
```


```{r}
#size factor
DESeq.gene.c <- estimateSizeFactors(DESeq.gene.c)
#normalize & log-sized
DESeq.gene_norm.c <- counts(DESeq.gene.c, normalized = TRUE)

```

Writing the gene expression for cyber sort
```{r, eval = FALSE}

#Ensembl to genename
gene_norm <- data.frame(log2(DESeq.gene_norm.c+1))
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
gene_IDs <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol","gene_biotype"),
              values = rownames(gene_norm), mart= mart)
gene_IDs <- subset(gene_IDs, hgnc_symbol!="" & !is.na(hgnc_symbol) )
gene_norm$Gene <- rep("",nrow(gene_norm))
gene_norm[gene_IDs$ensembl_gene_id_version,"Gene"]<- gene_IDs[,2]
bindtable2 <- subset(gene_norm, Gene!="" & !is.na(Gene) )
bindtable2 <- subset(bindtable2, Gene %in% unique(bindtable2$Gene))
bindtable2 <- bindtable2[!duplicated(bindtable2$Gene),]
rownames(bindtable2) <- make.names(bindtable2$Gene)
bindtable2 <- bindtable2 %>% dplyr::select(Gene, everything())
write.table(bindtable2, "MSKCC_Expression_Norm.txt", sep = "\t", row.names = FALSE)

```

#### Gene DESeq

```{r}
#create the object
#Include RECIST in design
DESeq.gene <- DESeqDataSetFromMatrix(countData = as.matrix(gene_table),
                                   colData = meta_sub,
                                   design = ~ RECIST) 
```



#### Process count info on gene
```{r}
par(mar=c(11, 8.1, 4.1, 2.1))
#size factor
DESeq.gene <- estimateSizeFactors(DESeq.gene)
#normalize & log-sized
DESeq.gene_norm <- counts(DESeq.gene, normalized = TRUE)
boxplot(log2(DESeq.gene_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.gene,"log.norm.counts") <- log2(DESeq.gene_norm+1)
```



```{r}
#Mean SD and PCA plot
par(mar=c(11, 4.1, 4.1, 4.1))
msd <- vsn::meanSdPlot(DESeq.gene@assays@data@listData$log.norm.counts, ranks=FALSE, plot = FALSE)
msd$gg + ggtitle("Sequencing depth normalized log2(read counts)") + ylab("standard deviation")

#PCA
DESeq.gene.trans <-DESeqTransform(DESeq.gene)
plotPCA(DESeq.gene.trans,intgroup = "Patient",returnData = FALSE) + ggtitle("Gene PCA Plot Patient")
plotPCA(DESeq.gene.trans,intgroup = "RECIST",returnData = FALSE) + ggtitle("Gene PCA Plot RECIST")
plotPCA(DESeq.gene.trans,intgroup = "RECIST_3m",returnData = FALSE) + ggtitle("Gene PCA Plot RECIST_3m")
```
Save PCA Results in the meta
```{r}
#return PC1 and PC2 value
pcas <- plotPCA(DESeq.gene.trans,intgroup = "RECIST",returnData = TRUE)
#add to metadata
meta_sub <- cbind(meta_sub, pcas[rownames(meta_sub),which(colnames(pcas) %in% c("PC1","PC2"))])
```

#### Run Analysis on the Gene

```{r}
DESeq.gene <- DESeq(DESeq.gene)
```

Thresholds

```{r}
### significance levels
MIN_FC=2
MIN_L2FC=log2(MIN_FC)
MAX_QVAL=0.05

### output controls
N_TOP_GENES=100
```

DE table
```{r}
#filter genes that pass the thresholds, organize by padj, filter to get only top 100 genes
gene_res <- results(DESeq.gene, independentFiltering = TRUE) %>% as.data.frame(.) %>% tibble::rownames_to_column() %>% 
  as_tibble(., rowname = NA) %>%
  dplyr::select(everything()) %>%
  arrange(padj) %>%
  filter(log2FoldChange >= MIN_L2FC | log2FoldChange <= -MIN_L2FC)
rownames(gene_res) <- make.names(gene_res$rowname)
gene_res[1:20,]
```

Checking Validity with QQplot & PCA of the gene expression data itself

```{r}
#order by padj
gene_res <- gene_res[order(gene_res$padj),]
n_quant<- sort(-log10(seq(1, 1/nrow(gene_res), length.out = nrow(gene_res))))
gene_res["Normal_Quantile"] <- n_quant
ggplot(gene_res, aes(x = Normal_Quantile, y =padj)) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for Gene Results\nDesign=RECIST,Group=Pre-tx Only")
```

```{r, eval = FALSE}
ggsave("Gene QQ Plot DE_Design=RECIST_Pre-tx.png")
```

```{r}
#scale the PC1 and PC2 value
meta_sub[,c(ncol(meta_sub)-1,ncol(meta_sub))] <- lapply(meta_sub[,c(ncol(meta_sub)-1,ncol(meta_sub))], function(x) c(scale(x)))
colnames(meta_sub)[c(ncol(meta_sub)-1,ncol(meta_sub))] <- c("Var1","Var2")

```

Heatmap

```{r}
#only do the top 100 
res_100 <- data.frame(gene_res) %>% head(n = N_TOP_GENES)
rownames(res_100) <- make.names(res_100$rowname)
#Ensembl to genename
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
gene_IDs <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol","gene_biotype"),
              values = res_100$rowname, mart= mart)
#replace row names with gene names
rownames(gene_IDs) <- make.names(gene_IDs$ensembl_gene_id_version)
res_100 <- merge(res_100,gene_IDs, by = 0, all = TRUE)
res_100 <-res_100[, -which(colnames(res_100) %in% c("Row.names"))]
res_100 <- res_100[order(res_100$padj),,drop = FALSE]
```


```{r}
#gene selection, use rlog scaled
DESeq.gene.rlog <- rlog(DESeq.gene, blind = TRUE)
#only take out the ones with gene names, and padj < 0.05
seg.gene <- subset(res_100, hgnc_symbol != "" & !is.na(hgnc_symbol) & padj < 0.05)$rowname
rlog.gene <-DESeq.gene.rlog[seg.gene,] %>% assay
```

```{r, fig.width = 8, fig.height = 6}
gene_label <- res_100[which(res_100$rowname %in% rownames(rlog.gene)), "hgnc_symbol"]
rownames(rlog.gene) <- gene_label
pheatmap(rlog.gene, scale="row",main = "DE genes (row-based z-score)\nDesign=RECIST,Pre-tx Only", fontsize = 10, size = 15, color = inferno(10), show_rownames = TRUE,cluster_rows = TRUE, legend_labels = "Expression Z-score")
```

```{r, eval = FALSE,fig.height = 10, fig.width = 8}
png("DE genes (z-score)_Design=RECIST_Pre-tx.png.png", res = 1200, unit = "in", width = 10, height = 15)
pheatmap(rlog.gene, scale="row", main = "DE genes (row-based z-score)\nDesign=RECIST,Pre-tx Only", fontsize = 11, size = 16, color = inferno(10), show_rownames = TRUE)

```


Volcano

```{r, fig.height=8}
all_IDs <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol","gene_biotype"),
              values = gene_res$rowname, mart= mart)
colnames(all_IDs) <- c("rowname","hgnc_symbol","gene_biotype")
gene_id <- merge(all_IDs, gene_res, on = "rowname" )
gene_id <- subset(gene_id, padj != "" & !is.na(padj))
EnhancedVolcano(gene_id, lab = as.character(gene_id$hgnc_symbol), x = 'log2FoldChange', y = 'padj', pCutoff = 0.01, labSize = 10,title = "Gene Vocalno, MSKCC Mets Pre-tx Samples\nDesign=RECIST,Padj cutoff=1*10^-2",titleLabSize = 40,legendLabSize = 20,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "grey50", "forestgreen", "red2"),pointSize = 4)
#ggsave( "Gene Vocalno_Design=RECIST_pre-tx.png")
```



### HERVs


```{r}
#create the object
DESeq.herv <- DESeqDataSetFromMatrix(countData = as.matrix(msk_sub),
                                   colData = meta_sub,
                                   design = ~ RECIST) 

```


#### Process count info 
```{r}
par(mar=c(9, 8, 4.1, 4.1))
#size factor
DESeq.herv <- estimateSizeFactors(DESeq.herv)
#normalize & log-sized
DESeq.herv_norm <- counts(DESeq.herv, normalized = TRUE)
boxplot(log2(DESeq.herv_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.herv,"log.norm.counts") <- log2(DESeq.herv_norm+1)
```


```{r}
#PCA
DESeq.herv.trans <-DESeqTransform(DESeq.herv)
plotPCA(DESeq.herv.trans,intgroup = "Patient",returnData = FALSE) + ggtitle("HERV PCA Plot\nColored by Patient")
plotPCA(DESeq.herv.trans,intgroup = "RECIST",returnData = FALSE) + ggtitle("HERV PCA Plot\nColored by RECIST")
plotPCA(DESeq.herv.trans,intgroup = "RECIST_3m",returnData = FALSE) + ggtitle("HERV PCA Plot\nColored by RECIST_3m")
```


```{r}
#has to specify that reference is "SD"
DESeq.herv$RECIST <- relevel(DESeq.herv$RECIST, ref = "SD")
DESeq.herv <- DESeq(DESeq.herv)
```


DE table on HERV
```{r}
#filter genes that pass the thresholds, organize by padj
herv_res <- results(DESeq.herv, independentFiltering = TRUE) %>% as.data.frame(.) %>% dplyr::select(everything()) %>%
  arrange(padj) %>%
  filter(log2FoldChange >= MIN_L2FC | log2FoldChange <= -MIN_L2FC)
herv_res[1:20,]
```

```{r, eval = FALSE}
write.csv(herv_res, "HERV_RESULTS_Design=RECIST.csv")

```
QQ Plot

```{r}
herv_res_all <- results(DESeq.herv) %>% as.data.frame()
Norm_quantile<- sort(-log10(seq(1, 1/nrow(herv_res_all), length.out = nrow(herv_res_all))))
herv_res_all <- herv_res_all[order(herv_res_all$padj, decreasing  = FALSE),]
herv_res_all["Normal_Quantile"] <- Norm_quantile
ggplot(herv_res_all, aes(x = Normal_Quantile, y =padj)) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for HERV\nDesign=RECIST")
```

```{r, eval = FALSE}
ggsave("HERV QQ Plot_Design=RECIST.png")
```

Heatmap

```{r}
#rlog scaled
DESeq.herv.rlog <- rlog(DESeq.herv, blind = TRUE)
#select padj < 0.05
seg.herv <- subset(herv_res, padj < 0.05 )
#kick out the L1F
herv_name <- str_extract(rownames(seg.herv), "^(ERV|HERV|LTR|MER|HML).+")

seg.herv1 <- seg.herv[(rownames(seg.herv) %in% herv_name),]
rlog.herv <-DESeq.herv.rlog[rownames(seg.herv1),] %>% assay

```



```{r,fig.height = 4, fig.width = 6}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(rlog.herv, scale="row",main = "DE HERVs (row-based z-score)\nDesign=RECIST, Pre-tx", fontsize = 15, size = 20, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score")
```

```{r, eval = FALSE,fig.height = 4, fig.width = 6}
png("DE HERVs Design=RECIST, Pre-tx.png", res = 1200, unit = "in", width = 8, height = 5.3)
pheatmap(rlog.herv, scale="row",main = "DE HERVs (row-based z-score)\nDesign=RECIST, Pre-tx", fontsize = 18, size = 25, color = inferno(25), show_rownames = TRUE, show_colnames = FALSE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score")
```

Volcano

```{r, fig.height = 7.5, fig.width = 6}
herv_res$hervs <- rownames(herv_res)
herv_res$hervs[(which(!(rownames(herv_res) %in% rownames(seg.herv1))))] <- ""
EnhancedVolcano(herv_res, lab = herv_res$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 0.05,title = "HERV Vocalno, MSKCC Mets, PD vs. SD, Pre-tx\nDesign=RECIST",labSize = 10,titleLabSize = 40,legendLabSize = 20,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-20,15)+theme( plot.title=element_text(size=30,face="bold"),axis.text=element_text(size=20),axis.title=element_text(size=20,face="bold"))
#ggsave("Valcano HERVs_Design=RECIST_PDvsSD.png")
```

Heatmap on Selective HERVs

```{r, fig.height = 5, fig.width = 1.6}
sd_sample <- rownames(meta_sub)[which(meta_sub$RECIST == "SD")]
#only take out the profile hERV expression
rlog.herv.profile <-DESeq.herv.rlog[herv_profile,sd_sample] %>% assay
#length(herv_profile)
pheatmap(rlog.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx SD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```
```{r, eval = FALSE,fig.height = 5, fig.width = 1.8}
png("Selected HERV Profile Expression Met-um SD.png", res = 1200, unit = "in", width = 3.6, height = 8)
pheatmap(rlog.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx SD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score", show_colnames = FALSE)
```


```{r, fig.height = 5, fig.width = 2}
pd_sample <- rownames(meta_sub)[which(meta_sub$RECIST == "PD")]
#only take out the profile hERV expression
rlog.herv.profile <-DESeq.herv.rlog[herv_profile,pd_sample] %>% assay
#length(herv_profile)
pheatmap(rlog.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx PD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```



```{r, eval = FALSE,fig.height = 5, fig.width = 2}
png("Selected HERV Profile Expression Met-um PD.png", res = 1200, unit = "in", width = 4, height = 8)
pheatmap(rlog.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx PD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score", show_colnames = FALSE)
```

# Individual Gene Count Plotting

Function to create error bars
```{r}
data_sum <- function(df, variable_row, grouping_row){
  g_factors <- as.factor(unique(df[,grouping_row]))
  sum_table <- data.frame(matrix(nrow = length(g_factors), ncol = 2))
  colnames(sum_table) <- c(variable_row, "sd")
  rownames(sum_table) <- g_factors
  for(i in 1:length(g_factors)){
    sub_mean <- mean(df[which(df[,grouping_row] == g_factors[i]),variable_row])
    sub_sd <- sd(df[which(df[,grouping_row] == g_factors[i]),variable_row])
    sum_table[i,c(1:2)] <- c(sub_mean,sub_sd)
  }
  sum_table[grouping_row] <- g_factors
  return(sum_table)
}

```


Significant count graphing 

```{r}
ERV316A3_6p21.33c <- plotCounts(DESeq.herv, gene = "ERV316A3_6p21.33c", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
ERV316A3_6p21.33c$RECIST <- relevel(ERV316A3_6p21.33c$RECIST, ref = "SD")
ERV316_summary<- data_sum(ERV316A3_6p21.33c, variable_row="count", grouping_row="RECIST")
ggplot(ERV316A3_6p21.33c, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for ERV316A3_6p21.33c\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data = ERV316_summary, width = 0.08, col = "grey50") + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data =ERV316_summary, width = 0.08, col = "red2")
#ggsave("ERV316A3_6p21.33c Counts.png")
```
```{r}
ERV316A3_3q13.12c <- plotCounts(DESeq.herv, gene = "ERV316A3_3q13.12c", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
ERV316A3_6p21.33c$RECIST <- relevel(ERV316A3_6p21.33c$RECIST, ref = "SD")
ERV3_summary<- data_sum(ERV316A3_3q13.12c, variable_row="count", grouping_row="RECIST")
ggplot(ERV316A3_3q13.12c, aes(x = RECIST, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for ERV316A3_3q13.12c\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data = ERV3_summary, width = 0.08, col = "grey50") + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = ERV3_summary, width = 0.08, col = "red2") +  theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))
#ggsave("ERV316A3_3q13.12c Counts.png")
```


```{r}
HML3_5p15.33c <- plotCounts(DESeq.herv, gene = "HML3_5p15.33c", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
HML3_5p15.33c$RECIST  <- relevel(HML3_5p15.33c$RECIST, ref = "SD")
HML3_summary<- data_sum(HML3_5p15.33c, variable_row="count", grouping_row="RECIST")
ggplot(HML3_5p15.33c, aes(x = RECIST, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for HML3_5p15.33c\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data = HML3_summary, width = 0.1, col = "grey50") + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = HML3_summary, width = 0.1, col = "red2")+theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))
#ggsave("HML3_5p15.33c Counts.png")
```


```{r}
ERVLB4_6q23.2c <- plotCounts(DESeq.herv, gene = "ERVLB4_6q23.2c", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
ERVLB_summary<- data_sum(ERVLB4_6q23.2c, variable_row="count", grouping_row="RECIST")
ggplot(ERVLB4_6q23.2c, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for ERVLB4_6q23.2c\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =ERVLB_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = ERVLB_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("ERVLB4_6q23.2c1 Counts.png")
```

```{r}
HERVW_2p23.1a <- plotCounts(DESeq.herv, gene = "HERVW_2p23.1a", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
HERVW_summary<- data_sum(HERVW_2p23.1a, variable_row="count", grouping_row="RECIST")
ggplot(HERVW_2p23.1a, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for HERVW_9p13.3\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =HERVW_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = HERVW_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("HERVW_2p23.1a Counts.png")
```




```{r}
PABLB_1q31.3a <- plotCounts(DESeq.herv, gene = "PABLB_1q31.3a", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
PABLB_summary<- data_sum(PABLB_1q31.3a, variable_row="count", grouping_row="RECIST")
ggplot(PABLB_1q31.3a, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for PABLB_1q31.3a\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data = PABLB_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = PABLB_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("PABLB_1q31.3a Counts.png")
HML3_19q13.31 <- plotCounts(DESeq.herv, gene = "HML3_19q13.31", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
HML3_19q_summary<- data_sum(HML3_19q13.31, variable_row="count", grouping_row="RECIST")
ggplot(HML3_19q13.31, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for HML3_19q13.31\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =HML3_19q_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = HML3_19q_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("HML3_19q13.31 Counts.png")
LTR19_12p13.31 <- plotCounts(DESeq.herv, gene = "LTR19_12p13.31", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
LTR_summary<- data_sum(LTR19_12p13.31, variable_row="count", grouping_row="RECIST")
ggplot(LTR19_12p13.31, aes(x = RECIST, y = count)) +geom_point() +  ggtitle("Normalized Counts for LTR19_12p13.31\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =LTR_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = LTR_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("LTR19_12p13.31 Counts.png")
```


### Plot Cybersort Results

```{r}
#load matrix
cyber_res <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/MSKCC_CIBERSORTx_Results_Absolute.csv", row.names = 1)
corr_sig <- cyber_res[,23:26]
fraction_table <- t(cyber_res[,1:22])
```
Only take out the ones matching primary paper

```{r}
cell_list <- c("T.cells.CD8","T.cells.follicular.helper","T.cells.CD4.naive","T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","B.cells.memory","B.cells.naive")
#order hervs by treatment group, then PD/SD
herv_reorder <- c(
  #pre & SD
  "SK_MEL_1316A_T"	,"SK_MEL_1313A_T","SK_MEL_1306A_T","SK_MEL_1331A_T",
  #pre & PD
  "SK_MEL_1299A_T","SK_MEL_1058B_T",	"SK_MEL_1330A_T","SK_MEL_1328A_T","SK_MEL_1327A_T",
  #On & SD
  "SK_MEL_1313B_T","SK_MEL_1306B_T","SK_MEL_1331B_T",
  #On & PD
  "SK_MEL_1299B_T","SK_MEL_1330B_T",
  #Post-PD
  "SK_MEL_1327B_T")
frac_sub <- fraction_table[cell_list,herv_reorder]
col_annotation <- sample_meta[herv_reorder,"Timepoint", drop = FALSE]
```


```{r,fig.height = 2, fig.width = 5}
pheatmap(frac_sub, scale='none',
show_rownames = TRUE,main = "Immune Cell Landscape\nMetastatic UM",color = inferno(10),fontsize = 15, size = 18, cluster_rows = FALSE, cluster_cols = FALSE, gaps_col = c(9,14), show_colnames = FALSE, annotation_col =col_annotation)

```

```{r, eval = FALSE, fig.height = 2, fig.width = 5}
png("Cybersort Results.png",res = 1200, unit = "in", width = 10, height = 4)
pheatmap(frac_sub, scale='none',
show_rownames = TRUE,main = "Immune Cell Landscape\nMetastatic UM",color = inferno(10),fontsize = 15, size = 18, cluster_rows = FALSE, cluster_cols = FALSE, gaps_col = c(9,14), show_colnames = FALSE, annotation_col =col_annotation)
```

### Primary vs. Mets


Read in the Cluster information from Bendall et al.
```{r}
load("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/primary_UM/analysis2/09-de.Rdata")
primary.cols <- data.frame(colData(tform))
```



create and combine metadata, rearrage so that the primary separates out by HC clusters, and the mets separates by SD, PD

```{r}
primary.cols <- primary.cols[order(primary.cols$clust.herv),]
counts.herv <- counts.herv[,rownames(primary.cols)]
```

```{r}
#metadata table
um_sub <- submeta(status = "primary", met_meta = meta_sub, count_table = counts.herv)
um_sub$hc_clust <- c(primary.cols[,"clust.herv"], rep("Metastatic", nrow(meta_sub)))
```


Combine the Reports
```{r}
combined_report <- cbind(counts.herv, msk_sub[rownames(counts.herv),])
#take out rows with NAs
combined_report <- na.omit(combined_report)
```

Run Primary vs. Metastatic

```{r}
#create the object
DESeq.herv.pm <- DESeqDataSetFromMatrix(countData = as.matrix(combined_report),
                                   colData = um_sub,
                                   design = ~ status) 

```


#### Process count info 
```{r, fig.width = 10, fig.height = 5}
#size factor
DESeq.herv.pm <- estimateSizeFactors(DESeq.herv.pm)
#normalize & log-sized
DESeq.herv.pm_norm <- counts(DESeq.herv.pm, normalized = TRUE)
boxplot(log2(DESeq.herv.pm_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.herv.pm,"log.norm.counts") <- log2(DESeq.herv.pm_norm+1)

DESeq.herv.pm$status <- relevel(DESeq.herv.pm$status, ref = "primary")
```


```{r}
#PCA
DESeq.herv.pm.trans <-DESeqTransform(DESeq.herv.pm)
plotPCA(DESeq.herv.pm.trans,intgroup = "status",returnData = FALSE) + ggtitle("HERV PCA Plot\nColored by Patient")
plotPCA(DESeq.herv.pm.trans,intgroup = "hc_clust",returnData = FALSE) + ggtitle("HERV PCA Plot") +xlim(-1*10^4.5,1*10^4.5)+ylim(-1*10^4.5,1*10^4.5)
#plotPCA(DESeq.herv.pm.trans,intgroup = "Patient",returnData = FALSE) + ggtitle("HERV PCA Plot") #+xlim(-1*10^4.5,1*10^4.5)+ylim(-1*10^4.5,1*10^4.5)
#need to correct for batch effect

pm.batch <- plotPCA(DESeq.herv.pm.trans,intgroup = "status",returnData = TRUE)
#normalize PC1 
pm.batch$PC1 <- scale(pm.batch$PC1)
pm.batch$PC2 <- scale(pm.batch$PC2)
pm.batch <- cbind(um_sub, pm.batch[rownames(um_sub), -which(colnames(pm.batch) %in% c("status", "group"))])
```

```{r}

#create the object
DESeq.herv.pm <- DESeqDataSetFromMatrix(countData = as.matrix(combined_report),
                                   colData =pm.batch,
                                   #use PC2 as correction
                                   design = ~ PC2 + hc_clust) 

DESeq.herv.pm <- estimateSizeFactors(DESeq.herv.pm)
#normalize & log-sized
DESeq.herv.pm_norm <- counts(DESeq.herv.pm, normalized = TRUE)

assay(DESeq.herv.pm,"log.norm.counts") <- log2(DESeq.herv.pm_norm+1)

DESeq.herv.pm$hc_clust <- relevel(DESeq.herv.pm$hc_clust, ref = "1")
```

```{r}
DESeq.herv.pm <- DESeq(DESeq.herv.pm)
```


DE table on HERV

```{r}
herv_res_pm<- restable(min_fc = 2, max_q = 0.05, n_top_genes = 20, DESeq.herv.pm)
```

```{r, eval = FALSE}
write.csv(herv_res_pm, "HERV_RESULTS_PvsM.csv")

```
QQ Plot

```{r}
all_res <- results(DESeq.herv.pm, independentFiltering = TRUE ) %>% as.data.frame(.)
Norm_quantile<- sort(-log10(seq(1, 1/nrow(all_res), length.out = nrow(all_res))))
all_res <- all_res[order(all_res$padj, decreasing  = FALSE),]
all_res["Normal_Quantile"] <- Norm_quantile
ggplot(all_res, aes(x = Normal_Quantile, y =padj)) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for HERV\nPrimary vs. Metastatic")
```

Heatmap

```{r}
pm_heat <- heattable(DESeq.herv.pm, herv_res_pm, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 3, max_q = 1*10^-6) 
 
```

```{r,fig.height = 10, fig.width = 12}
annotation_cols <- um_sub
annotation_cols$Status <- c(um_sub$hc_clust[1:80], sample_meta[rownames(um_sub)[81:nrow(um_sub)],"RECIST"])
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(pm_heat, scale="row",main = "DE HERVs, All Primary vs. Pre-tx Metastatic\nCutoff p= 1e-6, FoldChange = +/-3",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = annotation_cols[,"Status", drop = FALSE], gaps_col = c(20, 43, 61, 80))
```

```{r, eval = FALSE,fig.height = 20, fig.width = 15}
png("DE HERVs Design Primary vs Metastatic.png", res = 1200, unit = "in", width = 15, height = 20)

pheatmap(pm_heat, scale="row",main = "DE HERVs, All Primary vs. Pre-tx Metastatic\nCutoff p= 1e-6, FoldChange = +/-3",fontsize = 22, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = annotation_cols[,"Status", drop = FALSE], gaps_col = c(20, 43, 61, 80),  show_colnames = FALSE)

```

Volcano

```{r, fig.height = 10, fig.width = 15}
herv_res_pm$hervs <- rownames(herv_res_pm)
herv_res_pm$hervs[(which(!(rownames(herv_res_pm) %in% rownames(pm_heat))))] <- ""
EnhancedVolcano(herv_res_pm, lab = herv_res_pm$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 1*10^-6,FCcutoff = log2(3),title = "HERV Vocalno, Primary vs. Metastatic, p Cutoff = 1e-6, FC Cutoff = +/-3",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-10,17)+ylim(-0.1,20)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
ggsave("Valcano HERVs_PrimaryvsMets.png")
```

Count Plotting
```{r}

ERVLE_17p11.2c <- plotCounts(DESeq.herv.pm, gene = "ERVLE_17p11.2c", intgroup = "status", normalized = TRUE,returnData = TRUE)
ERVLE_17p11.2c$status <- relevel(ERVLE_17p11.2c$status, ref = "primary")
ERV_summary<- data_sum(ERVLE_17p11.2c, variable_row="count", grouping_row="status")
ggplot(ERVLE_17p11.2c, aes(x = status, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for ERVLE_17p11.2c\nSeparated by cancer status")+geom_errorbar(aes(x = status, ymin = 0, ymax = count+sd),data =ERV_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = status, ymin = count, ymax = count),data = ERV_summary, width = 0.08, col = "red2",colAlpha = 1/2)+theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))

#ggsave("ERVLE_17p11.2c.png")
```



```{r}

HML3_7q36.3 <- plotCounts(DESeq.herv.pm, gene = "HML3_7q36.3", intgroup = "status", normalized = TRUE,returnData = TRUE)
HML_summary<- data_sum(HML3_7q36.3, variable_row="count", grouping_row="status")
ggplot(HML3_7q36.3, aes(x = status, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for HML3_7q36.3\nSeparated by cancer status")+geom_errorbar(aes(x = status, ymin = 0, ymax = count+sd),data =HML_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = status, ymin = count, ymax = count),data = HML_summary, width = 0.08, col = "red2",colAlpha = 1/2)+theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))

ggsave("HML3_7q36.3.png")
```
```{r}

HML2_11q23.3 <- plotCounts(DESeq.herv.pm, gene = "HML2_11q23.3", intgroup = "status", normalized = TRUE,returnData = TRUE)
HML2_summary<- data_sum(HML2_11q23.3, variable_row="count", grouping_row="status")
ggplot(HML2_11q23.3, aes(x = status, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for HML2_11q23.3\nSeparated by cancer status")+geom_errorbar(aes(x = status, ymin = 0, ymax = count+sd),data =HML2_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = status, ymin = count, ymax = count),data = HML2_summary, width = 0.08, col = "red2",colAlpha = 1/2)+theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))

ggsave("HML2_11q23.3.png")
```

# Conduct the DESeq for Mets vs. Each Diagnostic Cluster

HC1
```{r}
HC1 <- rownames(primary.cols)[which(primary.cols$clust.herv == "C1")]
#metadata table
H1_sub <- submeta(status = "Primary_HC1", met_meta = sample_meta, count_table = counts.herv[,HC1])
HC1_report<- cbind(counts.herv[,HC1], msk_report[rownames(counts.herv),]) %>% na.omit(.)
```

DESeq
```{r}
DESeq.herv.HC1 <- DESeqDataSetFromMatrix(countData = as.matrix(HC1_report),
                                   colData = H1_sub,
                                   design = ~ status) 

```


#### Process count info 
```{r}

DESeq.herv.HC1 <- estimateSizeFactors(DESeq.herv.HC1)
#normalize & log-sized
DESeq.herv.HC1_norm <- counts(DESeq.herv.HC1, normalized = TRUE)

assay(DESeq.herv.HC1,"log.norm.counts") <- log2(DESeq.herv.HC1_norm+1)

DESeq.herv.HC1$status <- relevel(DESeq.herv.HC1$status, ref = "Primary_HC1")
```

```{r}
DESeq.herv.HC1 <- DESeq(DESeq.herv.HC1)
```


DE table on HERV

```{r}
hc1_res<- restable(min_fc = 3, max_q = 0.05, n_top_genes = 20, DESeq.herv.HC1)
```

Heatmap

```{r}
hc1_heat <- heattable(DESeq.herv.HC1, hc1_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 3, max_q = 1*10^-10) 
 
```


```{r,fig.height = 15, fig.width = 12}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(hc1_heat, scale="row",main = "DE HERVs, Primary HC1 vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = H1_sub[,"status", drop = FALSE], gaps_col = 20)
```

Volcano

```{r, fig.height = 10, fig.width = 15}
hc1_res$hervs <- rownames(hc1_res)
hc1_res$hervs[(which(!(rownames(hc1_res) %in% rownames(hc1_heat))))] <- ""
EnhancedVolcano(hc1_res, lab = hc1_res$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 1*10^-10,FCcutoff = log2(3),title = "HERV Vocalno, Primary HC1 vs. Metastatic, p Cutoff = 1e-10, FC Cutoff = +/-3",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-5, 15)+ylim(-0.1,20)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
#ggsave("Valcano HERVs_PrimaryvsMets.png")
```

HC2
```{r}
HC2 <- rownames(primary.cols)[which(primary.cols$clust.herv == "C2")]
#metadata table
H2_sub <- submeta(status = "Primary_HC2", met_meta = sample_meta, count_table = counts.herv[,HC2])
HC2_report<- cbind(counts.herv[,HC2], msk_report[rownames(counts.herv),]) %>% na.omit(.)
```

DESeq
```{r}
DESeq.herv.HC2 <- DESeqDataSetFromMatrix(countData = as.matrix(HC2_report),
                                   colData = H2_sub,
                                   design = ~ status) 

```


#### Process count info 
```{r}

DESeq.herv.HC2 <- estimateSizeFactors(DESeq.herv.HC2)
#normalize & log-sized
DESeq.herv.HC2_norm <- counts(DESeq.herv.HC2, normalized = TRUE)

assay(DESeq.herv.HC2,"log.norm.counts") <- log2(DESeq.herv.HC2_norm+1)

DESeq.herv.HC2$status <- relevel(DESeq.herv.HC2$status, ref = "Primary_HC2")
```

```{r}
DESeq.herv.HC2 <- DESeq(DESeq.herv.HC2)
```


DE table on HERV

```{r}
hc2_res<- restable(min_fc = 3, max_q = 0.05, n_top_genes = 20, DESeq.herv.HC2)
```

Heatmap

```{r}
hc2_heat <- heattable(DESeq.herv.HC2, hc2_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 3.5, max_q = 1*10^-12) 
 
```


```{r,fig.height = 20, fig.width = 12}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(hc2_heat, scale="row",main = "DE HERVs, Primary HC2 vs. Pre-tx Metastatic\nCutoff p= 1e-12, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = H2_sub[,"status", drop = FALSE], gaps_col = 23)
```

Volcano

```{r, fig.height = 10, fig.width = 15}
hc2_res$hervs <- rownames(hc2_res)
hc2_res$hervs[(which(!(rownames(hc2_res) %in% rownames(hc2_heat))))] <- ""
EnhancedVolcano(hc2_res, lab = hc2_res$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 1*10^-12,FCcutoff = log2(3.5),title = "HERV Vocalno, Primary HC2 vs. Metastatic, p Cutoff = 1e-12, FC Cutoff = +/-3.5",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-5,15)+ylim(-0.1,30)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
#ggsave("Valcano HERVs_PrimaryvsMets.png")
```
HC3

```{r}
HC3 <- rownames(primary.cols)[which(primary.cols$clust.herv == "C3")]
#metadata table
H3_sub <- submeta(status = "Primary_HC3", met_meta = sample_meta, count_table = counts.herv[,HC3])
HC3_report<- cbind(counts.herv[,HC3], msk_report[rownames(counts.herv),]) %>% na.omit(.)
```

DESeq
```{r}
DESeq.herv.HC3 <- DESeqDataSetFromMatrix(countData = as.matrix(HC3_report),
                                   colData = H3_sub,
                                   design = ~ status) 

```


#### Process count info 
```{r}

DESeq.herv.HC3 <- estimateSizeFactors(DESeq.herv.HC3)
#normalize & log-sized
DESeq.herv.HC3_norm <- counts(DESeq.herv.HC3, normalized = TRUE)

assay(DESeq.herv.HC3,"log.norm.counts") <- log2(DESeq.herv.HC3_norm+1)

DESeq.herv.HC3$status <- relevel(DESeq.herv.HC3$status, ref = "Primary_HC3")
```

```{r}
DESeq.herv.HC3 <- DESeq(DESeq.herv.HC3)
```


DE table on HERV

```{r}
hc3_res<- restable(min_fc = 3, max_q = 0.05, n_top_genes = 20, DESeq.herv.HC3)
```

Heatmap

```{r}
hc3_heat <- heattable(DESeq.herv.HC3, hc3_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 3, max_q = 1*10^-10) 
 
```


```{r,fig.height = 15, fig.width = 12}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(hc3_heat, scale="row",main = "DE HERVs, Primary HC3 vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = H3_sub[,"status", drop = FALSE], gaps_col = 18)
```

HC4

```{r}
HC4 <- rownames(primary.cols)[which(primary.cols$clust.herv == "C4")]
#metadata table
H4_sub <- submeta(status = "Primary_HC4", met_meta = sample_meta, count_table = counts.herv[,HC4])
HC4_report<- cbind(counts.herv[,HC4], msk_report[rownames(counts.herv),]) %>% na.omit(.)
```

DESeq
```{r}
DESeq.herv.HC4 <- DESeqDataSetFromMatrix(countData = as.matrix(HC4_report),
                                   colData = H4_sub,
                                   design = ~ status) 

```


#### Process count info 
```{r}

DESeq.herv.HC4 <- estimateSizeFactors(DESeq.herv.HC4)
#normalize & log-sized
DESeq.herv.HC4_norm <- counts(DESeq.herv.HC4, normalized = TRUE)

assay(DESeq.herv.HC4,"log.norm.counts") <- log2(DESeq.herv.HC4_norm+1)

DESeq.herv.HC4$status <- relevel(DESeq.herv.HC4$status, ref = "Primary_HC4")
```

```{r}
DESeq.herv.HC4 <- DESeq(DESeq.herv.HC4)
```


DE table on HERV

```{r}
hc4_res<- restable(min_fc = 3, max_q = 0.05, n_top_genes = 20, DESeq.herv.HC4)
```

Heatmap

```{r}
hc4_heat <- heattable(DESeq.herv.HC4, hc4_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 3, max_q = 1*10^-10) 
 
```

```{r,fig.height = 12, fig.width = 12}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(hc4_heat, scale="row",main = "DE HERVs, Primary HC4 vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = H4_sub[,"status", drop = FALSE], gaps_col = 19)
```





# Volcano Plots for Genes: Primary vs. Metastatic








# Functional Analysis
```{r}

library(ensembldb)
library(tidyr)
library(BSgenome)
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)

build_ens_DB<-function(){
  system("curl -q ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz --output Homo_sapiens.GRCh38.99.gtf.gz")
  DB<-ensembldb::ensDbFromGtf("Homo_sapiens.GRCh38.99.gtf.gz")
}
ifelse(!file.exists("Homo_sapiens.GRCh38.99.sqlite"),
       yes=build_ens_DB(), 
       no= "Annotations already downloaded")

```


```{r}
hg38_ens99<- ensembldb::EnsDb("Homo_sapiens.GRCh38.99.sqlite")

uscs_style <- mapSeqlevels(seqlevels(genes(hg38_ens99,filter=SeqNameFilter(c(1:22,"X","Y")))), "UCSC")
hg38_genes <- renameSeqlevels(genes(hg38_ens99,filter=SeqNameFilter(c(1:22,"X","Y")), columns=c("gene_name","gene_biotype")), uscs_style)
hg38_exons <- renameSeqlevels(exons(hg38_ens99, columns = c("tx_id", "tx_biotype", "gene_name", "gene_id"),filter =SeqNameFilter(c(1:22,"X","Y"))), uscs_style)

hg38_5utr <- renameSeqlevels(fiveUTRsByTranscript(hg38_ens99, columns = c("tx_id", "tx_biotype", "gene_name", "gene_id"),filter =SeqNameFilter(c(1:22,"X","Y"))), uscs_style)
hg38_3utr <- renameSeqlevels(threeUTRsByTranscript(hg38_ens99, columns = c("tx_id", "tx_biotype", "gene_name", "gene_id"),filter =SeqNameFilter(c(1:22,"X","Y"))), uscs_style)

build_tr_DB<-function(){
  system("curl -q --output HERV.gtf https://media.githubusercontent.com/media/mlbendall/telescope_annotation_db/master/builds/HERV_rmsk.hg38.v2/transcripts.gtf")
  system("curl -q --output L1.gtf https://media.githubusercontent.com/media/mlbendall/telescope_annotation_db/master/builds/L1Base.hg38.v1/transcripts.gtf")
  system("grep -p '\tgene\t' HERV.gtf | awk '{ split($10,locus,\"\\\"\");
class=\"HERV\"; split(locus[2],family,\"_\"); split($16,category,\"\\\"\");
chrom=$1;start=$4;end=$5;strand=$7;
print locus[2],class,family[1],category[2],chrom,start,end,strand;}' OFS='\t' >Retro_annotation.tsv")
  system(command = "grep -p '\texon\t' L1.gtf | awk '{ split($10,locus,\"\\\"\");
class=\"L1\"; family=\"L1\"; split($16,category,\"\\\"\");
chrom=$1;start=$4;end=$5;strand=$7;
print locus[2],class,family,category[2],chrom,start,end,strand;}' OFS='\t' >>Retro_annotation.tsv")
}
```


```{r}

ifelse(!file.exists("Retro_annotation.tsv"),
       yes=build_tr_DB(), 
       no= "Retro Annotations already downloaded")
hg38_retro<- read.table("Retro_annotation.tsv")
colnames(hg38_retro)<-c("Locus","Class","Family","Category","Chrom","Start","End","Strand")
rownames(hg38_retro)<-hg38_retro$Locus

herv_gr<-GRanges(seqnames = hg38_retro$Chrom, ranges = IRanges(start=hg38_retro$Start, end=hg38_retro$End), strand=hg38_retro$Strand)
names(herv_gr)<- hg38_retro$Locus


teseqs<- getSeq(BSgenome.Hsapiens.UCSC.hg38, herv_gr)
names(teseqs)<-paste0(names(herv_gr),"::",as.character(seqnames(herv_gr)),
                      ":",start(herv_gr),"-",end(herv_gr),"(",strand(herv_gr),")")
writeXStringSet(teseqs, "Retro.fa")
```

```{r}
cdpot<-read.delim("Coding_Potential_CNIT_TE.txt")
cdpot$ID<-sapply(strsplit(cdpot$Transcript.ID,split = ":"),'[[',1)
```

```{r}
hg38_retro$TE_CODING<- NA
hg38_retro[cdpot$ID,"TE_CODING"]<-cdpot$index

```