---
title: "Notes in Meta UM Proj"
author: "Phoebe Fei"
date: '2022-05-31'
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---

```{css style_settings, echo = FALSE}
blockquote { 
    font-size: 14px; 
    font-style: italic;
}
h1{
    font-size: 22px;
}
h2{
    font-size: 20px;
}
h3{
    font-size: 18px;
}
h4{
    font-size: 16px;
}

```

# Fastq to uBAM

## Notes:

Each sample is pair-end, and stored in /athena/nixonlab/scratch/projects/HERV_um_mets.git/raw. There are 5 fastq.gz files in each folder. *S.._R[1|2]_001.fastq.gz are the 2 files to use for downstream analysis.

```bach
nano qccheck.sh 

#! /bin/bash -l
#SBATCH --partition=panda
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --job-name=QC
#SBATCH --time=15:00:00 #HH/MM/SS
#SBATCH --mem=10G
spack load fastqc
cd /athena/nixonlab/scratch/projects/HERV_um_mets.git/raw
for d in */ ; do
    echo "$d"
    cd $d
    for f in *.fastq.gz; do
      fastqc $f -o /athena/nixonlab/scratch/tof4003/umprojscripts
    done
    cd ..
done 
cd /athena/nixonlab/scratch/tof4003/umprojscripts
spack load -r py-multiqc
multiqc .
```


### Libraries
```{r, message=FALSE}
library(stringr)
library(DESeq2)
library(vsn)
library(ggplot2)
library(apeglm)
library(dplyr)
library(magrittr)
library(pheatmap)
library(EnhancedVolcano)
library(viridis)
library(biomaRt)
library(PCAtools)
library(UpSetR)
library(edgeR)
library(scopetools)
```

### Now that there are samples, running DESeq2 on the Datasets

Sample names, organized by PD/SD
```{r}
samplenames <- c(
  #SD
  "SK_MEL_1316A_T"	,"SK_MEL_1313A_T","SK_MEL_1313B_T","SK_MEL_1306A_T","SK_MEL_1306B_T","SK_MEL_1331A_T","SK_MEL_1331B_T",	
  #PD
  "SK_MEL_1299A_T","SK_MEL_1299B_T","SK_MEL_1058B_T",	"SK_MEL_1330A_T","SK_MEL_1330B_T","SK_MEL_1328A_T",
  #PD (and has timepoint post-pd)
  "SK_MEL_1327A_T","SK_MEL_1327B_T")
```
Interested HERVs

```{r}
herv_profile <- c('ERVLE_17p11.2c','LTR46_Xq11.1','HERVH_6q24.1b','HERVH_3q21.3a','MER4B_3q21.2','HERVH_10q23.32','HML3_19q13.2','HERVI_9q22.1','MER4_1p21.3','HML3_3q21.2','ERVLE_5q13.2a','HERV3_19q13.42a','HERVE_Xp11.23','ERV316A3_6p21.33c','MER101_19q13.2c','ERVLE_2p13.2a','MER4B_12q22')

```

### Genes

#### Import readcounts and assign sample names
```{r}
genedirect <- "/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/align_multi/"
if(exists("bindtable")){
  rm("bindtable")}
for (i in 1:length(samplenames)){
  filedirect <- paste0(genedirect, samplenames[i],"/ReadsPerGene.out.tab")
  if(exists("bindtable")){
    test <- read.table(filedirect, header = FALSE, row.names = 1)
    #take out the summary rows
    sample.lane <- data.frame(test[-c(1:4),3], row.names = rownames(test)[-c(1:4)])
    bindtable <- cbind(bindtable,sample.lane[rownames(bindtable),])
  } else {
    test <- read.table(filedirect, header = FALSE, row.names = 1)
    #take out the summary rows
    bindtable <- data.frame(test[-c(1:4),3], row.names = rownames(test)[-c(1:4)])
  }
}
colnames(bindtable) <- samplenames
```

#### HERV counts


```{r, eval = FALSE}
devtools::install_github("nixonlab/scopetools")
```

```{r}
direct <- "/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/telescope/"
#create the directory path to reports.tsv
filedirect <- vector(length = length(samplenames))
for (i in 1:length(samplenames)){
  repo <- paste0(direct,samplenames[i], "/report.tsv")
  filedirect[i] <- repo
}
#make sample names as colnames
msk_report <- load_telescope_reports(files = filedirect, colnames = samplenames)

#take out no features 
msk_report <- msk_report[which(rownames(msk_report) != "__no_feature"),]
```


Take out the pre-treatment subgroup

```{r}
#make names rownames
rownames(sample_meta) <- make.names(sample_meta$SK.MEL)

meta_sub <- sample_meta[which(sample_meta$Timepoint == "Pre-tx"),]
meta_sub <- meta_sub[order(meta_sub$RECIST),]
gene_table <- bindtable[,meta_sub$SK.MEL]
msk_sub <- msk_report[,meta_sub$SK.MEL]
```

#### Gene + HERV to Plot PCA
```{r}
bind_all <- rbind(bindtable, msk_report[,colnames(msk_report)])
```

#### Primary Mets Data from TCGA

```{r}
load("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/primary_UM/analysis2/04-count_data.Rdata")
#data is the counts.herv and counts.tx
```

#### Gene Data
```{r}
load("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/primary_UM/analysis2/03-gene_data.Rdata")
#genes.annot
```

Combine primary + metastatic pre-tx colData
```{r}
#metadata table
col_info <- intersect(colnames(mdata), colnames(meta_sub))
um_sub <- rbind(mdata[,col_info, drop = FALSE], meta_sub[,col_info, drop = FALSE])
um_sub$Cluster <- c(mdata[,"Herv.cluster"], rep("Metastatic", nrow(meta_sub)))
um_sub$Class<- c(mdata[,"Herv.cluster"], meta_sub[,"RECIST"])
int_rows <- intersect(rownames(bind_all), rownames(counts.comb))
data_all <- cbind(bind_all[int_rows,], counts.comb[int_rows,])
#add a timepoint to mdata
mdata$Timepoint <- mdata$Herv.cluster
int_mdata <- intersect(colnames(mdata), colnames(sample_meta))
meta_all <- rbind(mdata[,int_mdata], sample_meta[,int_mdata])
```





## Functions

### Result Table & HeatMap Table Generation

Result Table
```{r}
restable <- function(min_fc = 2, max_p = 0.05, n_top_genes = 20, deseq, cont){
  MIN_L2FC=log2(min_fc)
  res_table <- results(deseq, independentFiltering = TRUE, contrast = cont) %>% as.data.frame(.) %>% dplyr::select(everything()) %>%
    filter(log2FoldChange >= min_fc | log2FoldChange <= -min_fc ) %>%
    filter(padj <= max_p) %>%
    arrange(padj) 
  print(res_table[1:n_top_genes, ])
  return(res_table)
}
```

Heatmap Table

```{r}
heattable <- function(deseq, res_table, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 2, max_p = 0.05 ){
  #rlog scaled
  deseq.vst <- varianceStabilizingTransformation(deseq, blind = TRUE)
  #select padj
  seg.herv <- subset(res_table, padj < max_p & (log2FoldChange > log2(min_fc) | log2FoldChange < -log2(min_fc)))
  #Select the HREVs/Genes
  herv_names  <- str_extract(rownames(seg.herv), select_string)
  seg.herv.1 <- seg.herv[(rownames(seg.herv) %in%  herv_names),]
  vst.herv <- deseq.vst[rownames(seg.herv.1),] %>% assay
  return(vst.herv)
}

```

### Cibersortx matrix

Outputs the Mixture file required by [Cibersortx](https://cibersortx.stanford.edu/):

**Mixture file format:**

*1.* Tab-delimited tabular input format (.txt) with no double quotations and no missing entries.

*2.* Genes in column 1; Mixture labels (sample names) in row 1

*3.* Given the significant difference between counts (e.g., CPM) and gene length-normalized expression data (e.g., TPM) we recommend that the signature matrix and mixture files be represented in the same normalization space whenever possible.

*4.* Data should be in non-log space. Note: if maximum expression value is less than 50; CIBERSORTx will assume that data are in log space, and will anti-log all expression values by 2x.

*5.* CIBERSORTx will add an unique identifier to each redundant gene symbol, however we recommend that users remove redundancy prior to file upload.

*6.* CIBERSORTx performs a feature selection and therefore typically does not use all genes in the signature matrix. It is generally ok if some genes are missing from the user’s mixture file. If less than 50% of signature matrix genes overlap, CIBERSORTx will issue a warning.

The default signature matrix to compute cell fraction on Cibersortx is [LM22](https://www.nature.com/articles/nmeth.3337#MOESM207) -  a validated leukocyte gene signature matrix that contains 547 genes distinguishing 22 human hematopoietic cell phenotypes, including seven T-cell types, naïve and memory B cells, plasma cells, natural killer (NK) cells and myeloid subsets. The expression matrix is RMA-normalized because they are from microarray results. 

The default normalization for the cibersortx function will be the median of ratios method of normalization implemented by the DESeq method. 

```{r}
library(dplyr)
library(magrittr)
library(DESeq2)
library(biomaRt)
library(edgeR)
library(DGEobj.utils)
#raw counts; methods = median of ratio, CPM, FPKM, FPK, or TPM; colnames order; reference table (if not provided, will connect to biomaRt),the coordinating names to capture and convert in the table (if using TPM, FPKM or FPK, have to provide the colname of exon length/gene length in the reference table)
cyber_out <- function(count_table, method = "median", name_sequence = colnames(count_table), ref_table = NA, name_cord = c("gene_id","gene_name","full_length")){
  count_table <- count_table[,name_sequence]
  if(is.na(ref_table)){
    #use the getBM method
    #Ensembl to genename
    mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
    ref_table <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol", "start_position","end_position"),values = rownames(count_table), mart= mart)
    #calculate the gene_length
    ref_table$full_length <- ref_table$end_position - ref_table$start_position
    colnames(ref_table) <- c("gene_id","gene_name","start","end","full_length")
  } else{
    ref_table <- ref_table[,name_cord]
    colnames(ref_table) <- c("gene_id","gene_name","full_length")
  }
  #normalization
  if(method %in% c("CPM","FPKM", "FPK" ,"TPM")){
    has_length <- intersect(rownames(count_table),ref_table$gene_id)
    #use convertCounts method from DGEobj.utils
    converted_mat <- convertCounts(
      countsMatrix = as.matrix(count_table[has_length,]),
      unit = method,
      geneLength = ref_table[has_length,"full_length"],
      log = FALSE,
      normalize = "none",
      prior.count = NULL
    )
  } else if ( method == "median"){
    #use DESeq default normalization method
    #create a pseudo colData (not important)
    counts_col <- data.frame(gene_ids = name_sequence)
    rownames(counts_col) <- name_sequence
    DESeq.c <- DESeqDataSetFromMatrix(countData = as.matrix(count_table),
                                      colData = counts_col,
                                      design = ~ gene_ids)
    #size factor
    DESeq.gene.c <- estimateSizeFactors(DESeq.c)
    #normalize & log-sized
    converted_mat <- counts(DESeq.gene.c, normalized = TRUE)
  }
  else{
    print("Method not supported. Returning original table")
    return(count_table)
  }
  #take out the na's
  ref_list_nona <- ref_table[which(ref_table$gene_id %in% rownames(count_table)), , drop = FALSE] %>% filter(., gene_name != "" & !is.na(gene_name))
  #rearrage based on provided colnames
  output_table <- converted_mat[ref_list_nona$gene_id,]
  rownames(output_table) <- ref_list_nona$gene_name
  colnames(output_table) <- name_sequence
  return(output_table)
}
```

```{r, eval = FALSE}
cyber_sample <-  c(
  #SD + Pre-tx
  "SK_MEL_1316A_T","SK_MEL_1313A_T","SK_MEL_1306A_T","SK_MEL_1331A_T",
  #SD + On-tx
  "SK_MEL_1313B_T","SK_MEL_1306B_T","SK_MEL_1331B_T",
  #PD + Pre-tx
  "SK_MEL_1299A_T","SK_MEL_1058B_T","SK_MEL_1327A_T",	"SK_MEL_1330A_T","SK_MEL_1328A_T",
  #PD + On-tx
  "SK_MEL_1299B_T","SK_MEL_1330B_T",
  #PD + Post-PD
  "SK_MEL_1327B_T")

ciber_tpm_msk <- cyber_out(count_table = bindtable, name_sequence = cyber_sample,method = "TPM", ref_table = genes.annot, name_cord = c("gene_id","gene_name","full_length" ))
ciber_msk_connect <- cyber_out(count_table = bindtable, name_sequence = cyber_sample)
ciber_msk <- cyber_out(count_table = bindtable, method = "median", name_sequence = cyber_sample, ref_table = genes.annot, name_cord = c("gene_id","gene_name","full_length" ))
ciber_primary <- cyber_out(count_table = counts.tx, method = "median", name_sequence = rownames(mdata), ref_table = genes.annot, name_cord = c("gene_id","gene_name","full_length" ))
ciber_primary_tpm <- cyber_out(count_table = counts.tx, name_sequence = rownames(mdata),method = "TPM", ref_table = genes.annot, name_cord = c("gene_id","gene_name","full_length" ))
ciber_primary_connect <- cyber_out(count_table = counts.tx, name_sequence = rownames(mdata))
```

```{r, eval = FALSE}
write.table(data.frame("Gene" = rownames(ciber_msk), ciber_msk), sep = "\t", "Median_Ratio_Normalized_Mixture_File_MSKCC.txt", row.names = F)

write.table(data.frame("Gene" = rownames(ciber_msk_connect), ciber_msk_connect), sep = "\t", "Median_Ratio_Normalized_Mixture_File_Biomart_Ref_MSKCC.txt", row.names = F)

write.table(data.frame("Gene" = rownames(ciber_tpm_msk), ciber_tpm_msk), sep = "\t","TPM_Normalized_Mixture_File_MSKCC.txt", row.names = F)
#note: will change all hyphens to dots and there is no way to reverse, gsub later
write.table(data.frame("Gene" = rownames(ciber_primary), ciber_primary), sep = "\t", "Median_Ratio_Normalized_Mixture_File_TCGA.txt", row.names = F)

write.table(data.frame("Gene" = rownames(ciber_primary_tpm), ciber_primary_tpm), sep = "\t","TPM_Normalized_Mixture_File_TCGA.txt",  row.names = F)

write.table(data.frame("Gene" = rownames(ciber_primary_connect), ciber_primary_connect), sep = "\t","Median_Ratio_Normalized_Mixture_File_Biomart_Ref_TCGA.txt" , row.names = F)
```

### Cibersortx Plotting

```{r}
#load matrix
ciber_tpm_primary <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_TPM_TCGA.csv", header = TRUE, row.names = 1)
ciber_tpm_msk <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_TPM_MSKCC.csv", header = TRUE, row.names = 1)
ciber_biomart_primary <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_TCGA_Biomart.csv", header = TRUE, row.names = 1)
ciber_biomart_msk <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_MSKCC_Biomart.csv", header = TRUE, row.names = 1)
ciber_msk <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_MSKCC.csv", header = TRUE, row.names = 1)
ciber_primary <- read.csv("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/Cybersort/CIBERSORTx_Results_TCGA.csv", header = TRUE, row.names = 1)
rownames(ciber_tpm_primary) <- gsub("\\.","-",rownames(ciber_tpm_primary))
rownames(ciber_biomart_primary) <- gsub("\\.","-",rownames(ciber_biomart_primary))
rownames(ciber_primary) <- gsub("\\.","-",rownames(ciber_primary))
rownames(ciber_tpm_msk) <- gsub("\\.","_",rownames(ciber_tpm_msk))
rownames(ciber_biomart_msk) <- gsub("\\.","_",rownames(ciber_biomart_msk))
rownames(ciber_msk) <- gsub("\\.","_",rownames(ciber_msk))
```


```{r}
ciber_heat <- function(prime_matrix, msk_matrix, title = "Immune Cell Infiltration Combined", col_split = NA, breaks = seq(0, 3, by = 0.25) , annotations = FALSE, show_colnames = FALSE, cluster_rows = FALSE, cluster_cols = FALSE, color = inferno(6), fontsize = 15, size = 18, reorder_list = c(colnames(prime_matrix), colnames(msk_matrix)), cell_str = "^(?!Mast|Eos|Dendritic).+"){
  #take out the significant matrix (p values, etc)
  corr_sig_p <- prime_matrix[,23:26]
  corr_sig_m <- msk_matrix[,23:26]
  print(rbind(corr_sig_p,  corr_sig_m[, colnames(corr_sig_p)]))
  fraction_table_primary <- t(prime_matrix[,1:22])
  fraction_table_msk<- t(msk_matrix[,1:22])
  cell_list <- which(!is.na(str_extract(rownames(fraction_table_primary), cell_str)))
  #combine the expression tables
  fraction_table_combined <- cbind(fraction_table_primary[cell_list,],fraction_table_msk[cell_list,])
  #reorder the expression table
  fraction_table_combined <- fraction_table_combined[,reorder_list]
  #graphing
  pheatmap(fraction_table_combined, scale='none',show_rownames = TRUE,main = title,color = color, breaks = breaks, fontsize = fontsize, size = size, cluster_rows = cluster_rows, cluster_cols = cluster_cols, gaps_col = col_split, show_colnames = show_colnames, annotation_col =annotations)
}
```

```{r, fig.height = 5, fig.width = 8}
herv_reorder <- c(
  #pre & SD
  "SK_MEL_1316A_T"	,"SK_MEL_1313A_T","SK_MEL_1306A_T","SK_MEL_1331A_T",
  #pre & PD
  "SK_MEL_1299A_T","SK_MEL_1058B_T",	"SK_MEL_1330A_T","SK_MEL_1328A_T","SK_MEL_1327A_T",
  #On & SD
  "SK_MEL_1313B_T","SK_MEL_1306B_T","SK_MEL_1331B_T",
  #On & PD
  "SK_MEL_1299B_T","SK_MEL_1330B_T",
  #Post-PD
  "SK_MEL_1327B_T")
primary_cols <- rownames(mdata)
col_annotation <- meta_all[c(primary_cols, herv_reorder),"Timepoint", drop = FALSE]
ciber_heat(prime_matrix = ciber_tpm_primary, msk_matrix = ciber_tpm_msk, title = "Immune Cell Infiltration,TPM Normalized", col_split = c(20, 43, 61, 80, 89, 94), breaks = seq(0, 2, by = 0.1) , annotations = col_annotation, show_colnames = TRUE, cluster_rows = FALSE, cluster_cols = FALSE, color = inferno(20), fontsize = 14, size = 20, reorder_list = c(primary_cols, herv_reorder), cell_str = "^(?!Mast|Eos|Dendritic).+")
```

```{r, fig.height = 5, fig.width = 8}
ciber_heat(prime_matrix = ciber_biomart_primary, msk_matrix = ciber_biomart_msk, title = "Immune Cell Infiltration, db.org Reference", col_split = c(20, 43, 61, 80, 89, 94), breaks = seq(0, 2, by = 0.1) , annotations = col_annotation, show_colnames = TRUE, cluster_rows = FALSE, cluster_cols = FALSE, color = inferno(20), fontsize = 14, size = 20, reorder_list = c(primary_cols, herv_reorder), cell_str = "^(?!Mast|Eos|Dendritic).+")
```


```{r, fig.height = 5, fig.width = 8}
ciber_heat(prime_matrix = ciber_primary, msk_matrix = ciber_msk, title = "Immune Cell Infiltration, Gene.annot Reference", col_split = c(20, 43, 61, 80, 89, 94), breaks = seq(0, 2.5, by = 0.1) , annotations = col_annotation, show_colnames = TRUE, cluster_rows = FALSE, cluster_cols = FALSE, color = inferno(25), fontsize = 14, size = 20, reorder_list = c(primary_cols, herv_reorder), cell_str = "^(?!Mast|Eos|Dendritic).+")
```


### DESeq Objects

#### HERV+Gene in All samples PCA Plotting

```{r}
DESeq.all <- DESeqDataSetFromMatrix(countData = as.matrix(data_all[,rownames(meta_all)]),
                                   colData = meta_all,
                                   design = ~ Timepoint) 

```

Mean SD Plot + PCA
```{r}
par(mar=c(11, 8.1, 4.1, 2.1))
#size factor
DESeq.all <- estimateSizeFactors(DESeq.all)
#normalize & log-sized
DESeq.all_norm <- counts(DESeq.all, normalized = TRUE)
boxplot(log2(DESeq.all_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.all,"log.norm.counts") <- log2(DESeq.all_norm+1)

#SD and PCA plot
msd.all <- vsn::meanSdPlot(DESeq.all@assays@data@listData$log.norm.counts, ranks=FALSE, plot = FALSE)
msd.all$gg + ggtitle("Sequencing depth normalized log2(read counts)") + ylab("standard deviation")
```

```{r}
#PCA
DESeq.all.trans <- DESeq(DESeq.all)
all.vst <- assay(vst(DESeq.all.trans))
all.pca <- pca(all.vst, metadata = colData(DESeq.all), removeVar = 0.1)
```

PCA Visualize

```{r}
screeplot(all.pca, axisLabSize = 18, titleLabSize = 22) + ggtitle("SCREE Plot for Gene + HERVs")
```
```{r}
biplot(all.pca, lab = NULL, colby = "Timepoint", legendPosition = 'right') + ggtitle("PCA, Gene+HERV\nColor by Patient Groups")
biplot(all.pca, lab = NULL, colby = "Age", legendPosition = 'right') + ggtitle("PCA, Gene+HERV\nColor by Age")
biplot(all.pca, lab = NULL, colby = "BAP1",legendPosition = 'right') + ggtitle("PCA, Gene+HERV\nColor by BAP1 Status")
biplot(all.pca, lab = NULL, colby = "Sex",legendPosition = 'right') + ggtitle("PCA, Gene+HERV\nColor by Sex")
```

### Primary vs. Mets


Combine the Reports
```{r}
load("/Users/phoebefei/Library/CloudStorage/OneDrive-med.cornell.edu/WCM/Metastatic UM Prj/primary_UM/analysis2/05-filter_counts.Rdata")
int_herv <- intersect(rownames(mfilt.herv), rownames(msk_sub))
combined_report <- cbind(mfilt.herv[int_herv,], msk_sub[int_herv,])
```

Run Primary vs. Metastatic

```{r}
#create the object
DESeq.herv.pm <- DESeqDataSetFromMatrix(countData = as.matrix(combined_report[,rownames(um_sub)]),
                                   colData = um_sub,
                                   design = ~ Cluster + 0) 

```


#### Process count info 
```{r, fig.width = 10, fig.height = 6}
#size factor
DESeq.herv.pm <- estimateSizeFactors(DESeq.herv.pm)
#normalize & log-sized
DESeq.herv.pm_norm <- counts(DESeq.herv.pm, normalized = TRUE)
boxplot(log2(DESeq.herv.pm_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.herv.pm,"log.norm.counts") <- log2(DESeq.herv.pm_norm+1)
```


```{r}
#SD and PCA plot
msd.herv.pm <- vsn::meanSdPlot(DESeq.herv.pm@assays@data@listData$log.norm.counts, ranks=FALSE, plot = FALSE)
msd.herv.pm$gg + ggtitle("Sequencing depth normalized log2(read counts)") + ylab("standard deviation")
DESeq.herv.pm.trans <- DESeq(DESeq.herv.pm)
herv.pm.vst <- assay(varianceStabilizingTransformation(DESeq.herv.pm.trans))
herv.pm.pca <- pca(herv.pm.vst, metadata = colData(DESeq.herv.pm.trans), removeVar = 0.1)
```

PCA Visualize

```{r}
screeplot(herv.pm.pca, axisLabSize = 18, titleLabSize = 22) + ggtitle("SCREE Plot for HERVs, Primary + Metastatic")
biplot(herv.pm.pca, lab = NULL, colby = "Cluster",legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by Cluster")
biplot(herv.pm.pca, lab = NULL, colby = "Class", legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by Primary and Mets Classificqtion")
biplot(herv.pm.pca, lab = NULL, colby = "Cancer_stage", legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by Cancer stage")
biplot(herv.pm.pca, lab = NULL, colby = "Age", legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by Age")
biplot(herv.pm.pca, lab = NULL, colby = "Sex", legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by Sex")
biplot(herv.pm.pca, lab = NULL, colby = "BAP1",legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by BAP1 Status")
biplot(herv.pm.pca, lab = NULL, colby = "SF3B1",legendPosition = 'right') + ggtitle("PCA, HERVs Primary + Mets\nColor by SF3B1 Status")
```
PC 1 separates Clusters 1;4;metastatic and clusters 2;3. Plot loadings to see the determinant variables.
```{r, fig.height = 8, fig.width = 4}
plotloadings(herv.pm.pca,
    components = getComponents(herv.pm.pca, c(1,2)),
    rangeRetain = 0.1,
    labSize = 4.0,
    title = 'Loadings plot for primary + metastatic HERVs',
    subtitle = 'PC1, PC2',
    caption = 'Top 1% variables',
    shape = 24,
    col = c('limegreen', 'black', 'red3'),
    drawConnectors = TRUE)
   #max.overlap = Inf

```

```{r}
#DESeq.herv.pm$Cluster <- relevel(DESeq.herv.pm$Cluster, ref = "Metastatic")
DESeq.herv.pm <- DESeq(DESeq.herv.pm,parallel = T)
```

DE table on HERV

```{r}
#resultsNames(DESeq.herv.pm)
#"Cluster1"          "Cluster2"          "Cluster3"          "Cluster4"          "ClusterMetastatic"
#Mets vs. All
herv_res_pm<-restable(min_fc = 3.5, max_p = 1e-10, n_top_genes = 20, DESeq.herv.pm, cont =c(-1/4,-1/4,-1/4,-1/4,+1))
#Mets vs. C1 vs.C4
herv_res_1_2<-restable(min_fc = 3.5, max_p = 1e-5, n_top_genes = 20, DESeq.herv.pm, cont =c(-1/2,0,0,-1/2,1))
#Mets vs. C1
herv_res_1<-restable(min_fc = 3.5, max_p = 1e-5, n_top_genes = 20, DESeq.herv.pm, cont = c("Cluster","1","Metastatic"))
#Mets vs. C2
herv_res_2<-restable(min_fc = 3.5, max_p = 1e-10, n_top_genes = 20, DESeq.herv.pm, cont = c("Cluster","2","Metastatic"))
#Mets vs. C3
herv_res_3<-restable(min_fc = 3.5, max_p = 1e-10, n_top_genes = 20, DESeq.herv.pm, cont = c("Cluster","3","Metastatic"))
#Mets vs. C4
herv_res_4<-restable(min_fc = 3.5, max_p = 1e-5, n_top_genes = 20, DESeq.herv.pm, cont = c("Cluster","4","Metastatic"))
```

```{r, eval = FALSE}
write.csv(herv_res_pm, "HERV_RESULTS_PvsM.csv")
```
QQ Plot

```{r}
herv_res_pm_q <- results(DESeq.herv.pm) %>% as.data.frame()
Norm_quantile<- sort(-log10(seq(1, 1/nrow(herv_res_pm_q), length.out = nrow(herv_res_pm_q))), decreasing = TRUE)
herv_res_pm_q <- herv_res_pm_q[order(herv_res_pm_q$padj, decreasing  = FALSE),]
herv_res_pm_q["Normal_Quantile"] <- Norm_quantile
ggplot(herv_res_pm_q, aes(x = Normal_Quantile, y =-log10(padj))) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for HERV Primary vs Mets\nDesign=RECIST")
```

Heatmap

```{r}
pm_heat <- heattable(DESeq.herv.pm, herv_res_pm, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-10) 
pm_1_2<-heattable(DESeq.herv.pm, herv_res_1_2, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-5)
pm_1<-heattable(DESeq.herv.pm, herv_res_1, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-5)
pm_2<-heattable(DESeq.herv.pm, herv_res_2, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-10)
pm_3<-heattable(DESeq.herv.pm, herv_res_3, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-10)
pm_4<-heattable(DESeq.herv.pm, herv_res_4, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc =3.5, max_p = 1e-5)
```

```{r,fig.height = 15, fig.width = 10}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(pm_heat, scale="row",main = "DE HERVs, All Primary vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE], gaps_col = c(20, 43, 61, 80))
pheatmap(pm_1[,c(which(um_sub$Cluster == "1"), which(um_sub$Cluster == "Metastatic"))], scale="row",main = "DE HERVs, C1 vs. Pre-tx Metastatic\nCutoff p= 1e-5, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE], gaps_col = 20)
pheatmap(pm_2[,c(which(um_sub$Cluster == "2"), which(um_sub$Cluster == "Metastatic"))], scale="row",main = "DE HERVs, C2 vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE])
pheatmap(pm_3[,c(which(um_sub$Cluster == "3"), which(um_sub$Cluster == "Metastatic"))], scale="row",main = "DE HERVs, C3 vs. Pre-tx Metastatic\nCutoff p= 1e-10, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE])
pheatmap(pm_4[,c(which(um_sub$Cluster == "4"), which(um_sub$Cluster == "Metastatic"))], scale="row",main = "DE HERVs, C4 vs. Pre-tx Metastatic\nCutoff p= 1e-5, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE])
pheatmap(pm_1_2[,c(which(um_sub$Cluster == "1"),which(um_sub$Cluster == "4"), which(um_sub$Cluster == "Metastatic"))], scale="row",main = "DE HERVs, C1 vs. C4 vs. Pre-tx Metastatic\nCutoff p= 1e-5, FoldChange = +/-3.5",fontsize = 20, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = um_sub[,"Cluster", drop = FALSE])
```

```{r, eval = FALSE,fig.height = 20, fig.width = 15}
png("DE HERVs Design Primary vs Metastatic.png", res = 1200, unit = "in", width = 15, height = 20)

pheatmap(pm_heat, scale="row",main = "DE HERVs, All Primary vs. Pre-tx Metastatic\nCutoff p= 1e-6, FoldChange = +/-3",fontsize = 22, size = 35, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = annotation_cols[,"Status", drop = FALSE], gaps_col = c(20, 43, 61, 80),  show_colnames = FALSE)

```

Volcano

```{r, fig.height = 10, fig.width = 15}
herv_res_all <- restable(min_fc = 2, max_p = 0.05, n_top_genes = 0, DESeq.herv.pm, cont =c(-1/4,-1/4,-1/4,-1/4,+1))
herv_res_all$hervs <- rownames(herv_res_all)
herv_res_all$hervs[(which(!(rownames(herv_res_all) %in% rownames(pm_heat))))] <- ""
EnhancedVolcano(herv_res_all, lab = herv_res_all$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 1*10^-10,FCcutoff = log2(3.5),title = "HERV Vocalno, Primary vs. Metastatic, p Cutoff = 1e-10, FC Cutoff = +/-3.5",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-10,17)+ylim(-0.1,20)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
#ggsave("Valcano HERVs_PrimaryvsMets.png")
```

```{r, fig.height = 10, fig.width = 15}
herv_res_14 <- restable(min_fc = 2, max_p = 0.05, n_top_genes = 0, DESeq.herv.pm, cont =c(-1/2,0,0,-1/2,+1))
herv_res_14$hervs <- rownames(herv_res_14)
herv_res_14$hervs[(which(!(rownames(herv_res_14) %in% rownames(pm_heat))))] <- ""
EnhancedVolcano(herv_res_14, lab = herv_res_14$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 1*10^-5,FCcutoff = log2(3.5),title = "HERV Vocalno, C1 vs. C4 vs. Metastatic, p Cutoff = 1e-5, FC Cutoff = +/-3.5",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-10,17)+ylim(-0.1,20)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
#ggsave("Valcano HERVs_PrimaryvsMets.png")
```

# Volcano Plots for Genes: Primary vs. Metastatic


counts combined
```{r}
combined_gene<- cbind(bindtable[rownames(mfilt.tx),rownames(meta_sub)], mfilt.tx)
#take out rows with NAs
combined_gene <- na.omit(combined_gene)
```

Run Primary vs. Metastatic

```{r}
#create the object
DESeq.gene.pm <- DESeqDataSetFromMatrix(countData = as.matrix(combined_gene[,rownames(um_sub)]),
                                   colData = um_sub,
                                   design = ~ Cluster + 0) 

```


#### Process count info 
```{r, fig.width = 10, fig.height = 6}
#size factor
DESeq.gene.pm <- estimateSizeFactors(DESeq.gene.pm)
#normalize & log-sized
DESeq.gene.pm_norm <- counts(DESeq.gene.pm, normalized = TRUE)
boxplot(log2(DESeq.gene.pm_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.gene.pm,"log.norm.counts") <- log2(DESeq.gene.pm_norm+1)
```


```{r}
#SD and PCA plot
msd.gene.pm <- vsn::meanSdPlot(DESeq.gene.pm@assays@data@listData$log.norm.counts, ranks=FALSE, plot = FALSE)
msd.gene.pm$gg + ggtitle("Sequencing depth normalized log2(read counts)") + ylab("standard deviation")
DESeq.gene.pm.trans <- DESeq(DESeq.gene.pm)
gene.pm.vst <- assay(varianceStabilizingTransformation(DESeq.gene.pm.trans))
gene.pm.pca <- pca(gene.pm.vst, metadata = colData(DESeq.gene.pm.trans), removeVar = 0.3)
```

PCA Visualize

```{r}
screeplot(gene.pm.pca, axisLabSize = 18, titleLabSize = 22) + ggtitle("SCREE Plot for Genes, Primary + Metastatic")
biplot(gene.pm.pca, lab = NULL, colby = "Cluster",legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by Cluster")
biplot(gene.pm.pca, lab = NULL, colby = "Class", legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by Primary and Mets Classificqtion")
biplot(gene.pm.pca, lab = NULL, colby = "Cancer_stage", legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by Cancer stage")
biplot(gene.pm.pca, lab = NULL, colby = "Age", legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by Age")
biplot(gene.pm.pca, lab = NULL, colby = "Sex", legendPosition = 'right') + ggtitle("PCA, GenesPrimary + Mets\nColor by Sex")
biplot(gene.pm.pca, lab = NULL, colby = "BAP1",legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by BAP1 Status")
biplot(gene.pm.pca, lab = NULL, colby = "SF3B1",legendPosition = 'right') + ggtitle("PCA, Genes Primary + Mets\nColor by SF3B1 Status")
```
PC 1 separates Clusters 1;4;metastatic and clusters 2;3. Plot loadings to see the determinant variables.
```{r, fig.height = 8, fig.width = 4}
plotloadings(gene.pm.pca,
    components = getComponents(gene.pm.pca, c(1,2)),
    rangeRetain = 0.1,
    labSize = 4.0,
    title = 'Loadings plot for primary + metastatic Genes',
    subtitle = 'PC1, PC2',
    caption = 'Top 1% variables',
    shape = 24,
    col = c('limegreen', 'black', 'red3'),
    drawConnectors = TRUE)
   #max.overlap = Inf
```

```{r}
DESeq.gene.pm$Cluster <- relevel(DESeq.gene.pm$Cluster, ref = "Metastatic")
DESeq.gene.pm <- DESeq(DESeq.gene.pm,parallel = T)
```

DE table on HERV

```{r}
#resultsNames(DESeq.herv.pm)
# "Cluster1"          "Cluster2"          "Cluster3"          "Cluster4"          "ClusterMetastatic"
#Mets vs. All
gene_res_pm<-restable(min_fc = 3, max_p = 1e-15, n_top_genes = 20, DESeq.gene.pm, cont = c(-1/4,-1/4,-1/4,-1/4,+1))
#Mets vs. C1 vs. C4
gene_res_1_4<-restable(min_fc = 3, max_p = 1e-10, n_top_genes = 20, DESeq.gene.pm, cont =c(-1/2,0,0,-1/2,+1))
```


Volcano

```{r, fig.height = 10, fig.width = 15}
gene_res_all <- restable(min_fc = 2, max_p = 0.05, n_top_genes = 0, DESeq.gene.pm, cont =c(-1/4,-1/4,-1/4,-1/4,+1))
gene_res_all$gene <- rownames(gene_res_all)
gene_res_all$gene[(which(!(rownames(gene_res_all) %in% rownames(gene_res_pm))))] <- ""
#to actual gene name
gene_res_all$gene_name <- genes.annot[gene_res_all$gene, "gene_name"]
EnhancedVolcano(gene_res_all, lab = gene_res_all$gene_name, x = 'log2FoldChange', y = 'padj', pCutoff =  1e-15,FCcutoff = log2(3),title = "Gene Vocalno, Primary vs. Metastatic, p Cutoff = 1e-15, FC Cutoff = +/-3",titleLabSize = 30, labSize = 10,legendLabSize = 15,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-10,17)+ylim(-0.1,20)+theme( plot.title=element_text(size=45,face="bold"),axis.text=element_text(size=25),axis.title=element_text(size=25,face="bold"))
#ggsave("Valcano HERVs_PrimaryvsMets.png")
```


#### HERV & BAP1 Status



#### Individual Count Plotting

Function to create error bars
```{r}
data_sum <- function(df, variable_row, grouping_row){
  g_factors <- as.factor(unique(df[,grouping_row]))
  sum_table <- data.frame(matrix(nrow = length(g_factors), ncol = 2))
  colnames(sum_table) <- c(variable_row, "sd")
  rownames(sum_table) <- g_factors
  for(i in 1:length(g_factors)){
    sub_mean <- mean(df[which(df[,grouping_row] == g_factors[i]),variable_row])
    sub_sd <- sd(df[which(df[,grouping_row] == g_factors[i]),variable_row])
    sum_table[i,c(1:2)] <- c(sub_mean,sub_sd)
  }
  sum_table[grouping_row] <- g_factors
  return(sum_table)
}

```


Significant count graphing 


```{r}
ERV316A3_3q13.12c <- plotCounts(DESeq.herv.pm, gene = "ERV316A3_3q13.12c", intgroup = "Cluster", normalized = TRUE,returnData = TRUE)
ERV316A3_3q13.12c$Cluster <- relevel(ERV316A3_3q13.12c$Cluster, ref = "SD")
ERV3_summary<- data_sum(ERV316A3_3q13.12c, variable_row="count", grouping_row="Cluster")
ggplot(ERV316A3_3q13.12c, aes(x = Cluster, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for ERV316A3_3q13.12c\nSeparated by Cluster")+geom_errorbar(aes(x = Cluster, ymin = count-sd, ymax = count+sd),data = ERV3_summary, width = 0.08, col = "grey50") + geom_errorbar(aes(x = Cluster, ymin = count, ymax = count),data = ERV3_summary, width = 0.08, col = "red2") +  theme( plot.title=element_text(size=16,face="bold"),axis.text=element_text(size=18),axis.title=element_text(size=20,face="bold"))
#ggsave("ERV316A3_3q13.12c Counts.png")
```


```{r}
HML2_19q11 <- plotCounts(DESeq.herv.pm, gene = "HML2_19q11", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
HML2_19q11$RECIST <- relevel(HML2_19q11$RECIST, ref = "SD")
HML2_summary<- data_sum(HML2_19q11, variable_row="count", grouping_row="RECIST")
ggplot(HML2_19q11, aes(x = RECIST, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for HML2_19q11\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =HML2_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = HML2_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("ERVLB4_6q23.2c1 Counts.png")
```

```{r}
MER4_Xq21.33c <- plotCounts(DESeq.herv.pm, gene = "MER4_Xq21.33c", intgroup = "RECIST", normalized = TRUE,returnData = TRUE)
MER4_summary<- data_sum(MER4_Xq21.33c, variable_row="count", grouping_row="RECIST")
ggplot(MER4_Xq21.33c, aes(x = RECIST, y = count)) +geom_point(aes(size = 1.5)) +  ggtitle("Normalized Counts for MER4_Xq21.33c\nSeparated by RECIST Status")+geom_errorbar(aes(x = RECIST, ymin = count-sd, ymax = count+sd),data =MER4_summary, width = 0.08, col = "grey50",colAlpha = 1/2) + geom_errorbar(aes(x = RECIST, ymin = count, ymax = count),data = MER4_summary, width = 0.08, col = "red2",colAlpha = 1/2)
#ggsave("HERVW_2p23.1a Counts.png")
```


#### Reproduce the primary paper

```{r}
p.cutoff <- 1e-3
lfc.cutoff <- 1.5
countDat <- filt.herv
cat(sprintf('%d variables\n', nrow(countDat)))

stopifnot(all(colnames(countDat) == rownames(samples)))
#samples$clust.herv <- clust.df$clust.retro.k4

# Wald test
dds <- DESeq2::DESeqDataSetFromMatrix(countDat, colData(tform), ~clust.herv + 0)
dds <- DESeq2::DESeq(dds, parallel=T)
tform <- DESeq2::varianceStabilizingTransformation(dds, blind=FALSE)

# Extract results
res <- list(
  "C1" = results(dds, contrast=c(+1, -1/3, -1/3, -1/3), alpha=p.cutoff),
  "C2" = results(dds, contrast=c(-1/3, +1, -1/3, -1/3), alpha=p.cutoff),
  "C3" = results(dds, contrast=c(-1/3, -1/3, +1, -1/3), alpha=p.cutoff),
  "C4" = results(dds, contrast=c(-1/3, -1/3, -1/3, +1), alpha=p.cutoff),
  "C12vC34" = results(dds, contrast=c(+1/2, +1/2, -1/2, -1/2), alpha=p.cutoff),
  "C1vC2"   = results(dds, contrast=c(  +1,   -1,    0,    0), alpha=p.cutoff),
  "C3vC4"   = results(dds, contrast=c(   0,    0,   +1,   -1), alpha=p.cutoff)
)

res <- lapply(res, function(r) {
  r$class <- loc_class[rownames(r),]$class
  r$display <- loc_class[rownames(r),]$display
  r
})

sig <- lapply(res, function(r) {
  s <- subset(r, padj < p.cutoff & abs(log2FoldChange) > lfc.cutoff)
  s[order(s$padj),]
})
```

```{r, fig.height = 8, fig.width = 8}
C12v34heat <- heattable(dds, data.frame(sig$C12vC34) , select_string = ".+",min_fc = 3.5, max_q = 1e-10)
C12v34heat_ordered <- C12v34heat[,rownames(col_annotation_tcga)]
pheatmap(C12v34heat_ordered, scale="row",main = "HeatMap TCGA 1vs2vs3vs4", fontsize = 10, size = 15, color = inferno(25), breaks = breaklist, show_rownames = TRUE,cluster_rows = TRUE,cluster_cols = FALSE, legend_labels = "Expression Z-score", annotation_col = col_annotation_tcga)
```



#### Gene DESeq on Pre-tx Only

```{r}
#create the object
#Include RECIST in design
DESeq.gene <- DESeqDataSetFromMatrix(countData = as.matrix(gene_table),
                                   colData = meta_sub,
                                   design = ~ RECIST) 
```



#### Process count info on gene
```{r}
par(mar=c(11, 8.1, 4.1, 2.1))
#size factor
DESeq.gene <- estimateSizeFactors(DESeq.gene)
#normalize & log-sized
DESeq.gene_norm <- counts(DESeq.gene, normalized = TRUE)
boxplot(log2(DESeq.gene_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.gene,"log.norm.counts") <- log2(DESeq.gene_norm+1)
```



```{r}
#PCA
DESeq.gene.trans <- DESeq(DESeq.gene)
gene.vst <- assay(vst(DESeq.gene.trans))
gene.pca <- pca(gene.vst, metadata = colData(DESeq.gene), removeVar = 0.01)
```

PCA Visualize

```{r}
screeplot(gene.pca, axisLabSize = 18, titleLabSize = 22) + ggtitle("SCREE Plot for Genes")
biplot(gene.pca, lab = gene.pca$metadata$Patient, colby = "Patient") + ggtitle("PCA, Gene\nColor by Patients")
biplot(gene.pca, lab = NULL, colby = "RECIST", legendPosition = 'right') + ggtitle("PCA, Gene\nColor by RECIST Status")
biplot(gene.pca, lab = NULL, colby = "Age", legendPosition = 'right') + ggtitle("PCA, Gene\nColor by Age")
biplot(gene.pca, lab = NULL, colby = "BAP1",legendPosition = 'right') + ggtitle("PCA, Gene\nColor by BAP1 Status")
biplot(gene.pca, lab = NULL, colby = "Sex",legendPosition = 'right') + ggtitle("PCA, Gene\nColor by Sex")
biplot(gene.pca, lab = NULL, colby = "RECIST_3m", legendPosition = 'right') + ggtitle("PCA, Gene\nColor by RECIST after 3M of Tx")
biplot(gene.pca, lab = NULL, colby = "Stage", legendPosition = 'right') + ggtitle("PCA, Gene\nColor by Met Stage")
```

#### Run Analysis on the gene with proper design

```{r}
DESeq.gene <- DESeqDataSetFromMatrix(countData = as.matrix(gene_table),
                                   colData = meta_sub,
                                   design = ~ Sex + Age + RECIST + 0) 
```


Thresholds

```{r}
### significance levels
MIN_FC=2
MIN_L2FC=log2(MIN_FC)
MAX_QVAL=0.05

### output controls
N_TOP_GENES=100
```

```{r}
#has to specify that reference is "SD"
DESeq.gene$RECIST <- relevel(DESeq.gene$RECIST, ref = "SD")
DESeq.gene <- DESeq(DESeq.gene, parallel = TRUE)
```

DE table
```{r}
#filter genes that pass the thresholds, organize by padj, filter to get only top 100 genes
gene_res <- restable(min_fc = 2, max_q = 0.05, n_top_genes = N_TOP_GENES, DESeq.gene, cont = c("RECIST","PD","SD"))
gene_res[1:20,]
```

Checking Validity with QQplot & PCA of the gene expression data itself

```{r}
#order by padj
gene_res <- gene_res[order(gene_res$padj),]
n_quant<- sort(-log10(seq(1, 1/nrow(gene_res), length.out = nrow(gene_res))), decreasing = TRUE)
gene_res["Normal_Quantile"] <- n_quant
ggplot(gene_res, aes(x = Normal_Quantile, y =-log10(padj))) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for Gene Results\nDesign=RECIST,Group=Pre-tx Only")
```

```{r, eval = FALSE}
ggsave("Gene QQ Plot DE_Design=RECIST_Pre-tx.png")
```

Heatmap

```{r}
#only do the top 100 
res_100 <- data.frame(gene_res) %>% head(n = N_TOP_GENES)
res_100$rowname <- rownames(res_100)
#Ensembl to genename
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
gene_IDs <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol","gene_biotype"),
              values = res_100$rowname, mart= mart)
#replace row names with gene names
rownames(gene_IDs) <- make.names(gene_IDs$ensembl_gene_id_version)
res_100 <- merge(res_100,gene_IDs, by = 0, all = TRUE)
res_100 <-res_100[, -which(colnames(res_100) %in% c("Row.names"))]
res_100 <- res_100[order(res_100$padj),,drop = FALSE]

```


```{r}
vst.gene <- heattable(deseq = DESeq.gene, herv_res = gene_res, select_string = ".+",min_fc = 2, max_q = 0.05)
gene_label <- res_100[which(res_100$rowname %in% rownames(vst.gene)), "hgnc_symbol"]
gene_label[which(gene_label == "")] <- NA
vst_gene_sense <- vst.gene[!is.na(gene_label),]
rownames(vst_gene_sense) <- gene_label[!is.na(gene_label)]
```

```{r, fig.width = 8, fig.height = 6}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(vst_gene_sense, scale="row",main = "DE genes (row-based z-score)\nDesign=RECIST,Pre-tx Only", fontsize = 10, size = 15, color = inferno(25), breaks = breaklist, show_rownames = TRUE,cluster_rows = TRUE,cluster_cols = FALSE, legend_labels = "Expression Z-score", annotation_col = meta_sub[,"RECIST", drop = FALSE])
```

```{r, eval = FALSE,fig.height = 10, fig.width = 8}
png("DE genes (z-score)_Design=RECIST_Pre-tx.png.png", res = 1200, unit = "in", width = 10, height = 15)
pheatmap(vst_gene_sense, scale="row",main = "DE genes (row-based z-score)\nDesign=RECIST,Pre-tx Only", fontsize = 10, size = 15, color = inferno(25), breaks = breaklist, show_rownames = TRUE, show_colnames = FALSE, cluster_rows = TRUE, legend_labels = "Expression Z-score")
```


Volcano

```{r, fig.height=8}
all_IDs <- getBM(filters= "ensembl_gene_id_version", attributes= c("ensembl_gene_id_version","hgnc_symbol","gene_biotype"),values = rownames(gene_res), mart= mart)
colnames(all_IDs) <- c("rowname","hgnc_symbol","gene_biotype")
gene_res$rowname <- rownames(gene_res)
gene_id <- merge(all_IDs, gene_res, on = "rowname" )
EnhancedVolcano(gene_id, lab = as.character(gene_id$hgnc_symbol), x = 'log2FoldChange', y = 'padj', pCutoff = 0.05, FCcutoff = log2(2),labSize = 10,title = "Gene Vocalno, MSKCC Mets Pre-tx Samples\nDesign=RECIST+Sex+Age",titleLabSize = 40,legendLabSize = 20,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "grey50", "forestgreen", "red2"),pointSize = 4)+xlim(-20,20)+ylim(-0.01, 6)
#ggsave( "Gene Vocalno_Design=RECIST_pre-tx.png")
```



### HERVs: PD vs.SD


```{r}
#create the object
DESeq.herv <- DESeqDataSetFromMatrix(countData = as.matrix(msk_sub),
                                   colData = meta_sub,
                                   design = ~ Age+Sex+RECIST+0) 

```


#### Process count info 
```{r}
par(mar=c(9, 8, 4.1, 4.1))
#size factor
DESeq.herv <- estimateSizeFactors(DESeq.herv)
#normalize & log-sized
DESeq.herv_norm <- counts(DESeq.herv, normalized = TRUE)
boxplot(log2(DESeq.herv_norm+1), notch=FALSE, main = "Size-factor-normalized read counts", ylab="log2(read counts)", cex = .6, las = 2)
assay(DESeq.herv,"log.norm.counts") <- log2(DESeq.herv_norm+1)
```


```{r}
#PCA
DESeq.herv.trans <- DESeq(DESeq.herv)
herv.vst <- assay(vst(DESeq.herv.trans))
herv.pca <- pca(herv.vst, metadata = colData(DESeq.herv), removeVar = 0.1)
```

PCA Visualize

```{r}
screeplot(herv.pca, axisLabSize = 18, titleLabSize = 22) + ggtitle("SCREE Plot for HERVs")
biplot(herv.pca, lab = NULL, colby = "RECIST", legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by RECIST Status")
biplot(herv.pca, lab = NULL, colby = "Age", legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by Age")
biplot(herv.pca, lab = NULL, colby = "BAP1",legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by BAP1 Status")
biplot(herv.pca, lab = NULL, colby = "SF3B1",legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by SF3B1 Status")
biplot(herv.pca, lab = NULL, colby = "Sex",legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by Sex")
biplot(herv.pca, lab = NULL, colby = "RECIST_3m", legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by RECIST after 3M of Tx")
biplot(herv.pca, lab = NULL, colby = "Stage", legendPosition = 'right') + ggtitle("PCA, HERVs\nColor by Met Stage")
```

```{r}
DESeq.herv$RECIST <- relevel(DESeq.herv$RECIST, ref = "SD")
DESeq.herv <- DESeq(DESeq.herv)
```

DE table on HERV
```{r}
#filter genes that pass the thresholds, organize by padj
herv_res <- restable(min_fc = 2, max_q = 0.05, n_top_genes = N_TOP_GENES, DESeq.herv, cont = c("RECIST","PD","SD"))
```

```{r, eval = FALSE}
write.csv(herv_res, "HERV_RESULTS_Design=RECIST.csv")

```
QQ Plot

```{r}
herv_res_all <- results(DESeq.herv) %>% as.data.frame()
Norm_quantile<- sort(-log10(seq(1, 1/nrow(herv_res_all), length.out = nrow(herv_res_all))), decreasing = TRUE)
herv_res_all <- herv_res_all[order(herv_res_all$padj, decreasing  = FALSE),]
herv_res_all["Normal_Quantile"] <- Norm_quantile
ggplot(herv_res_all, aes(x = Normal_Quantile, y =-log10(padj))) + geom_point()+geom_abline(intercept = 0, slope = 1, color="red") + xlab("Negative Log Normal Quantile (Expected)")+ylab("Negative Log 10 P Values (Observed)")+ggtitle("QQ Plot for HERV\nDesign=RECIST")
```

```{r, eval = FALSE}
ggsave("HERV QQ Plot_Design=RECIST.png")
```

Heatmap

```{r}
#rlog scaled
herv.vst <- heattable(DESeq.herv, herv_res, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 2, max_q = 0.5) 

```



```{r,fig.height = 3, fig.width = 6}
breaklist <- seq(-3, 3, by = 0.25)
pheatmap(herv.vst, scale="row",main = "DE HERVs (row-based z-score)\nDesign=RECIST, Pre-tx", fontsize = 15, size = 20, color = inferno(25), show_rownames = TRUE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = meta_sub[,"RECIST", drop = FALSE])
```

```{r, eval = FALSE,fig.height = 5, fig.width = 5.5}
png("DE HERVs Design=RECIST, Pre-tx.png", res = 1200, unit = "in", width = 6, height = 6)
pheatmap(herv.vst, scale="row",main = "DE HERVs (row-based z-score)\nDesign=RECIST, Pre-tx", fontsize = 15, size = 15, color = inferno(25), show_rownames = TRUE, show_colnames = FALSE,cluster_cols = FALSE, cluster_rows = TRUE, breaks = breaklist, legend_labels = "Expression Z-score", annotation_col = meta_sub[,"RECIST", drop = FALSE])
```

Volcano

```{r, fig.height = 6, fig.width = 6}
herv_res$hervs <- rownames(herv_res)
herv_res$hervs[(which(!(rownames(herv_res) %in% rownames(herv.vst))))] <- ""
EnhancedVolcano(herv_res, lab = herv_res$hervs, x = 'log2FoldChange', y = 'padj', pCutoff = 0.05,title = "HERV Vocalno, MSKCC Mets, PD vs. SD, Pre-tx\nDesign=RECIST",labSize = 10,titleLabSize = 40,legendLabSize = 20,max.overlaps = Inf, drawConnectors = TRUE,arrowheads = FALSE,colAlpha = 1/2,col = c("black", "forestgreen", "yellow", "red2"),pointSize = 4)+xlim(-20,15)+theme( plot.title=element_text(size=30,face="bold"),axis.text=element_text(size=20),axis.title=element_text(size=20,face="bold"))
#ggsave("Valcano HERVs_Design=RECIST_PDvsSD.png")
```

Heatmap on Selective HERVs

```{r, fig.height = 5, fig.width = 1.6}
sd_sample <- rownames(meta_sub)[which(meta_sub$RECIST == "SD")]
#only take out the profile hERV expression
herv.res.all <- restable(min_fc = 0, max_q = 1, n_top_genes = N_TOP_GENES, DESeq.herv, cont = c("RECIST","PD","SD"))
herv.vst.all <- heattable(DESeq.herv, herv.res.all, select_string = "^(ERV|HERV|LTR|MER|HML).+",min_fc = 0, max_q = 1) 
vst.herv.profile <-herv.vst.all[herv_profile,sd_sample]
#length(herv_profile)
pheatmap(vst.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx SD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```
```{r, eval = FALSE,fig.height = 5, fig.width = 1.8}
png("Selected HERV Profile Expression Met-um SD.png", res = 1200, unit = "in", width = 3.6, height = 8)
pheatmap(vst.herv.profile, scale="row",main = "17 HERV Profile\nPre-tx SD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, show_colnames = FALSE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```


```{r, fig.height = 5, fig.width = 2}
pd_sample <- rownames(meta_sub)[which(meta_sub$RECIST == "PD")]
#only take out the profile hERV expression
vst.herv.profile.pd <-herv.vst.all[herv_profile,pd_sample]
#length(herv_profile)
pheatmap(vst.herv.profile.pd, scale="row",main = "17 HERV Profile\nPre-tx PD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```



```{r, eval = FALSE,fig.height = 5, fig.width = 2}
png("Selected HERV Profile Expression Met-um PD.png", res = 1200, unit = "in", width = 4, height = 8)
pheatmap(vst.herv.profile.pd, scale="row",main = "17 HERV Profile\nPre-tx PD", fontsize = 10, size = 10, color = inferno(25), breaks = breaklist, show_rownames = TRUE, show_colnames = FALSE, cluster_rows = FALSE, cluster_cols = TRUE,legend_labels = "Expression Z-score")
```

